// tldraw/dist-cjs/lib/shapes/note/toolStates/Pointing.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Pointing_exports = {};
__export(Pointing_exports, {
  Pointing: () => Pointing,
  createNoteShape: () => createNoteShape,
  getNoteShapeAdjacentPositionOffset: () => getNoteShapeAdjacentPositionOffset
});
module.exports = __toCommonJS(Pointing_exports);
var import_editor = require("@tldraw/editor");
var import_selectHelpers = require("../../../tools/SelectTool/selectHelpers");
var import_getDisplayValues = require("../../shared/getDisplayValues");
var import_noteHelpers = require("../noteHelpers");
class Pointing extends import_editor.StateNode {
  static id = "pointing";
  dragged = false;
  info = {};
  markId = "";
  shape = {};
  shapeId;
  shapeCenter = new import_editor.Vec();
  hasCreatedShape = false;
  onEnter() {
    const { editor } = this;
    this.hasCreatedShape = false;
    this.shapeId = (0, import_editor.createShapeId)();
    this.markId = editor.markHistoryStoppingPoint(`creating_note:${this.shapeId}`);
    const noteUtil = editor.getShapeUtil("note");
    const dv = (0, import_getDisplayValues.getDisplayValues)(noteUtil, { props: noteUtil.getDefaultProps() });
    const center = editor.inputs.getOriginPagePoint().clone();
    const offset = getNoteShapeAdjacentPositionOffset(
      editor,
      center,
      editor.getResizeScaleFactor(),
      dv.noteWidth,
      dv.noteHeight
    );
    if (offset) {
      center.sub(offset);
    }
    this.shapeCenter = center;
    if (!editor.getInstanceState().isCoarsePointer) {
      this.ensureShapeCreated();
    }
  }
  // Create the note now if it hasn't been created yet (idempotent). The deferred
  // drag and release paths route through here; a max-shapes failure cancels the
  // gesture, matching the original onEnter behaviour.
  ensureShapeCreated() {
    if (this.hasCreatedShape) return;
    const shape = createNoteShape(this.editor, this.shapeId, this.shapeCenter);
    if (shape) {
      this.shape = shape;
      this.hasCreatedShape = true;
    } else {
      this.cancel();
    }
  }
  onPointerMove(info) {
    if (this.editor.inputs.getIsDragging()) {
      this.ensureShapeCreated();
      if (!this.hasCreatedShape) return;
      this.editor.setCurrentTool("select.translating", {
        ...info,
        target: "shape",
        shape: this.shape,
        onInteractionEnd: "note",
        isCreating: true,
        creatingMarkId: this.markId,
        onCreate: () => {
          (0, import_selectHelpers.startEditingShapeWithRichText)(this.editor, this.shape.id);
        }
      });
    }
  }
  onPointerUp() {
    this.complete();
  }
  onLongPress() {
    (0, import_editor.cancelShapeCreationOnLongPress)(this.editor, () => this.cancel());
  }
  onInterrupt() {
    this.cancel();
  }
  onComplete() {
    this.complete();
  }
  onCancel() {
    this.cancel();
  }
  complete() {
    this.ensureShapeCreated();
    if (!this.hasCreatedShape) return;
    if (this.editor.getInstanceState().isToolLocked) {
      this.parent.transition("idle");
    } else {
      (0, import_selectHelpers.startEditingShapeWithRichText)(this.editor, this.shape.id, { info: this.info });
    }
  }
  cancel() {
    this.editor.bailToMark(this.markId);
    this.parent.transition("idle", this.info);
  }
}
function getNoteShapeAdjacentPositionOffset(editor, center, scale, noteWidth, noteHeight) {
  let min = import_noteHelpers.NOTE_ADJACENT_POSITION_SNAP_RADIUS / editor.getZoomLevel();
  let offset;
  for (const pit of (0, import_noteHelpers.getAvailableNoteAdjacentPositions)(editor, {
    rotation: 0,
    scale,
    extraHeight: 0,
    noteWidth,
    noteHeight
  })) {
    const deltaToPit = import_editor.Vec.Sub(center, pit);
    const dist = deltaToPit.len();
    if (dist < min) {
      min = dist;
      offset = deltaToPit;
    }
  }
  return offset;
}
function createNoteShape(editor, id, center) {
  editor.createShape({
    id,
    type: "note",
    x: center.x,
    y: center.y,
    props: {
      scale: editor.getResizeScaleFactor()
    }
  });
  const shape = editor.getShape(id);
  if (!shape) return;
  editor.select(id);
  const bounds = editor.getShapeGeometry(shape).bounds;
  const newPoint = (0, import_editor.maybeSnapToGrid)(
    new import_editor.Vec(shape.x - bounds.width / 2, shape.y - bounds.height / 2),
    editor
  );
  editor.updateShapes([
    {
      id,
      type: "note",
      x: newPoint.x,
      y: newPoint.y
    }
  ]);
  return editor.getShape(id);
}
