// tldraw/dist-cjs/lib/styles.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var styles_exports = {};
__export(styles_exports, {
  STYLES: () => STYLES,
  getColorStyleItems: () => getColorStyleItems,
  getFontStyleItems: () => getFontStyleItems
});
module.exports = __toCommonJS(styles_exports);
var import_editor = require("@tldraw/editor");
function isPaletteColor(value) {
  return typeof value === "object" && value !== null && "solid" in value;
}
function getColorStyleItems(colors) {
  const result = [];
  const seen = /* @__PURE__ */ new Set();
  for (const name of import_editor.DefaultColorStyle.values) {
    if (name in colors && isPaletteColor(colors[name])) {
      if (name === "white") continue;
      result.push({ value: name, icon: "color" });
      seen.add(name);
    }
  }
  for (const [key, value] of Object.entries(colors)) {
    if (!seen.has(key) && key !== "white" && isPaletteColor(value)) {
      result.push({ value: key, icon: "color" });
    }
  }
  return result;
}
const defaultFontIcons = {
  draw: "font-draw",
  sans: "font-sans",
  serif: "font-serif",
  mono: "font-mono"
};
function getFontStyleItems(theme) {
  const result = [];
  const seen = /* @__PURE__ */ new Set();
  for (const name of import_editor.DefaultFontStyle.values) {
    const entry = theme.fonts[name];
    if (name in theme.fonts && (0, import_editor.isFontEntry)(entry)) {
      result.push({ value: name, icon: fontIcon(entry, name) });
      seen.add(name);
    }
  }
  for (const [key, value] of Object.entries(theme.fonts)) {
    if (!seen.has(key) && (0, import_editor.isFontEntry)(value)) {
      result.push({ value: key, icon: fontIcon(value, key) });
    }
  }
  return result;
}
function fontIcon(font, name) {
  if (font.icon != null) return font.icon;
  return defaultFontIcons[name] ?? "font-draw";
}
const STYLES = {
  fill: [
    { value: "none", icon: "fill-none" },
    { value: "semi", icon: "fill-semi" },
    { value: "solid", icon: "fill-solid" }
  ],
  fillExtra: [
    { value: "pattern", icon: "fill-pattern" },
    { value: "lined-fill", icon: "fill-lined-fill" },
    { value: "fill", icon: "fill-fill" }
  ],
  dash: [
    { value: "draw", icon: "dash-draw" },
    { value: "dashed", icon: "dash-dashed" },
    { value: "dotted", icon: "dash-dotted" },
    { value: "solid", icon: "dash-solid" }
  ],
  size: [
    { value: "s", icon: "size-small" },
    { value: "m", icon: "size-medium" },
    { value: "l", icon: "size-large" },
    { value: "xl", icon: "size-extra-large" }
  ],
  font: [
    { value: "draw", icon: "font-draw" },
    { value: "sans", icon: "font-sans" },
    { value: "serif", icon: "font-serif" },
    { value: "mono", icon: "font-mono" }
  ],
  textAlign: [
    { value: "start", icon: "text-align-left" },
    { value: "middle", icon: "text-align-center" },
    { value: "end", icon: "text-align-right" }
  ],
  horizontalAlign: [
    { value: "start", icon: "horizontal-align-start" },
    { value: "middle", icon: "horizontal-align-middle" },
    { value: "end", icon: "horizontal-align-end" }
  ],
  verticalAlign: [
    { value: "start", icon: "vertical-align-start" },
    { value: "middle", icon: "vertical-align-middle" },
    { value: "end", icon: "vertical-align-end" }
  ],
  arrowKind: [
    { value: "arc", icon: "arrow-arc" },
    { value: "elbow", icon: "arrow-elbow" }
  ],
  arrowheadStart: [
    { value: "none", icon: "arrowhead-none" },
    { value: "arrow", icon: "arrowhead-arrow" },
    { value: "triangle", icon: "arrowhead-triangle" },
    { value: "square", icon: "arrowhead-square" },
    { value: "dot", icon: "arrowhead-dot" },
    { value: "diamond", icon: "arrowhead-diamond" },
    { value: "inverted", icon: "arrowhead-triangle-inverted" },
    { value: "bar", icon: "arrowhead-bar" }
  ],
  arrowheadEnd: [
    { value: "none", icon: "arrowhead-none" },
    { value: "arrow", icon: "arrowhead-arrow" },
    { value: "triangle", icon: "arrowhead-triangle" },
    { value: "square", icon: "arrowhead-square" },
    { value: "dot", icon: "arrowhead-dot" },
    { value: "diamond", icon: "arrowhead-diamond" },
    { value: "inverted", icon: "arrowhead-triangle-inverted" },
    { value: "bar", icon: "arrowhead-bar" }
  ],
  spline: [
    { value: "line", icon: "spline-line" },
    { value: "cubic", icon: "spline-cubic" }
  ]
};
