// tldraw/dist-cjs/lib/assets/ImageAssetUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ImageAssetUtil_exports = {};
__export(ImageAssetUtil_exports, {
  ImageAssetUtil: () => ImageAssetUtil
});
module.exports = __toCommonJS(ImageAssetUtil_exports);
var import_editor = require("@tldraw/editor");
var import_defaultExternalContentHandlers = require("../defaultExternalContentHandlers");
var import_assets = require("../utils/assets/assets");
class ImageAssetUtil extends import_editor.AssetUtil {
  static type = "image";
  static props = import_editor.imageAssetProps;
  static migrations = import_editor.imageAssetMigrations;
  options = {
    maxDimension: import_defaultExternalContentHandlers.DEFAULT_MAX_IMAGE_DIMENSION,
    supportedMimeTypes: null
  };
  getDefaultProps() {
    return {
      w: 0,
      h: 0,
      name: "",
      isAnimated: false,
      mimeType: null,
      src: null
    };
  }
  getSupportedMimeTypes() {
    return this.options.supportedMimeTypes ?? import_editor.DEFAULT_SUPPORTED_IMAGE_TYPES;
  }
  async getAssetFromFile(file, assetId) {
    const doc = this.editor.getContainerDocument();
    const size = await import_editor.MediaHelpers.getImageSize(file, doc);
    const isAnimated = await import_editor.MediaHelpers.isAnimated(file);
    const pixelRatio = "pixelRatio" in size && size.pixelRatio !== 1 ? size.pixelRatio : void 0;
    const assetInfo = {
      id: assetId,
      type: "image",
      typeName: "asset",
      props: {
        name: file.name,
        src: "",
        w: size.w,
        h: size.h,
        fileSize: file.size,
        mimeType: file.type,
        isAnimated,
        ...pixelRatio ? { pixelRatio } : void 0
      },
      meta: {}
    };
    const maxDimension = this.options.maxDimension;
    if (maxDimension && isFinite(maxDimension)) {
      const originalSize = { w: assetInfo.props.w, h: assetInfo.props.h };
      const resizedSize = (0, import_assets.containBoxSize)(originalSize, { w: maxDimension, h: maxDimension });
      if (originalSize !== resizedSize && import_editor.MediaHelpers.isStaticImageType(file.type)) {
        assetInfo.props.w = resizedSize.w;
        assetInfo.props.h = resizedSize.h;
      }
    }
    return assetInfo;
  }
}
