// tldraw/dist-cjs/lib/assets/VideoAssetUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var VideoAssetUtil_exports = {};
__export(VideoAssetUtil_exports, {
  VideoAssetUtil: () => VideoAssetUtil
});
module.exports = __toCommonJS(VideoAssetUtil_exports);
var import_editor = require("@tldraw/editor");
class VideoAssetUtil extends import_editor.AssetUtil {
  static type = "video";
  static props = import_editor.videoAssetProps;
  static migrations = import_editor.videoAssetMigrations;
  options = {
    supportedMimeTypes: null
  };
  getDefaultProps() {
    return {
      w: 0,
      h: 0,
      name: "",
      isAnimated: true,
      mimeType: null,
      src: null
    };
  }
  getSupportedMimeTypes() {
    return this.options.supportedMimeTypes ?? import_editor.DEFAULT_SUPPORT_VIDEO_TYPES;
  }
  async getAssetFromFile(file, assetId) {
    let fileType = file.type;
    if (fileType === "video/quicktime") {
      fileType = "video/mp4";
    }
    const doc = this.editor.getContainerDocument();
    const size = await import_editor.MediaHelpers.getVideoSize(file, doc);
    return {
      id: assetId,
      type: "video",
      typeName: "asset",
      props: {
        name: file.name,
        src: "",
        w: size.w,
        h: size.h,
        fileSize: file.size,
        mimeType: fileType,
        isAnimated: true
      },
      meta: {}
    };
  }
}
