// tldraw/dist-cjs/lib/ui/components/SharePanel/DefaultPeopleMenu.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DefaultPeopleMenu_exports = {};
__export(DefaultPeopleMenu_exports, {
  DefaultPeopleMenu: () => DefaultPeopleMenu
});
module.exports = __toCommonJS(DefaultPeopleMenu_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_editor = require("@tldraw/editor");
var import_radix_ui = require("radix-ui");
var import_components = require("../../context/components");
var import_useCollaborationStatus = require("../../hooks/useCollaborationStatus");
var import_useMenuIsOpen = require("../../hooks/useMenuIsOpen");
var import_useTranslation = require("../../hooks/useTranslation/useTranslation");
var import_OfflineIndicator = require("../OfflineIndicator/OfflineIndicator");
var import_DefaultPeopleMenuContent = require("./DefaultPeopleMenuContent");
function DefaultPeopleMenu({ children }) {
  const msg = (0, import_useTranslation.useTranslation)();
  const dir = (0, import_useTranslation.useDirection)();
  const container = (0, import_editor.useContainer)();
  const editor = (0, import_editor.useEditor)();
  const userIds = (0, import_editor.usePeerIds)();
  const userColor = (0, import_editor.useValue)("user", () => editor.user.getColor(), [editor]);
  const userName = (0, import_editor.useValue)("user", () => editor.user.getName(), [editor]);
  const [isOpen, onOpenChange] = (0, import_useMenuIsOpen.useMenuIsOpen)("people menu");
  const collaborationStatus = (0, import_useCollaborationStatus.useCollaborationStatus)();
  const { PeopleMenuFacePile } = (0, import_components.useTldrawUiComponents)();
  if (collaborationStatus === "offline") {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_OfflineIndicator.OfflineIndicator, {});
  }
  if (!userIds.length || !children && !PeopleMenuFacePile) {
    return null;
  }
  const content = children ?? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DefaultPeopleMenuContent.DefaultPeopleMenuContent, { userIds });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_radix_ui.Popover.Root, { onOpenChange, open: isOpen, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_radix_ui.Popover.Trigger, { dir, asChild: true, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("button", { className: "tlui-people-menu__avatars-button", title: msg("people-menu.title"), children: PeopleMenuFacePile ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(PeopleMenuFacePile, { userColor, userIds, userName }) : null }) }),
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_radix_ui.Popover.Portal, { container, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_radix_ui.Popover.Content,
      {
        dir,
        className: "tlui-menu",
        side: "bottom",
        sideOffset: 2,
        collisionPadding: 4,
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "tlui-people-menu__wrapper", children: content })
      }
    ) })
  ] });
}
