// tldraw/dist-cjs/lib/overlays/BrushOverlayUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BrushOverlayUtil_exports = {};
__export(BrushOverlayUtil_exports, {
  BrushOverlayUtil: () => BrushOverlayUtil
});
module.exports = __toCommonJS(BrushOverlayUtil_exports);
var import_editor = require("@tldraw/editor");
class BrushOverlayUtil extends import_editor.OverlayUtil {
  static type = "brush";
  options = {
    zIndex: 300,
    getDefaultDisplayValues(editor, _overlay, theme, colorMode) {
      const colors = theme.colors[colorMode];
      return {
        fillColor: colors.brushFill,
        strokeColor: colors.brushStroke,
        lineWidth: 1
      };
    },
    getCustomDisplayValues() {
      return {};
    }
  };
  isActive() {
    return this.editor.getInstanceState().brush !== null;
  }
  getOverlays() {
    const brush = this.editor.getInstanceState().brush;
    if (!brush) return [];
    return [
      {
        id: "brush",
        type: "brush",
        props: {
          x: brush.x,
          y: brush.y,
          w: Math.max(1, brush.w),
          h: Math.max(1, brush.h)
        }
      }
    ];
  }
  render(ctx, overlays) {
    const overlay = overlays[0];
    if (!overlay) return;
    const { x, y, w, h } = overlay.props;
    const zoom = this.editor.getZoomLevel();
    const dv = (0, import_editor.getOverlayDisplayValues)(this, overlay);
    ctx.fillStyle = dv.fillColor;
    ctx.fillRect(x, y, w, h);
    ctx.lineWidth = dv.lineWidth / zoom;
    ctx.strokeStyle = dv.strokeColor;
    ctx.strokeRect(x, y, w, h);
  }
  renderMinimap(ctx, overlays, zoom) {
    const overlay = overlays[0];
    if (!overlay) return;
    const { x, y, w, h } = overlay.props;
    const dv = (0, import_editor.getOverlayDisplayValues)(this, overlay);
    ctx.fillStyle = dv.fillColor;
    ctx.fillRect(x, y, w, h);
    ctx.lineWidth = dv.lineWidth / zoom;
    ctx.strokeStyle = dv.strokeColor;
    ctx.strokeRect(x, y, w, h);
  }
}
