// tldraw/dist-cjs/lib/overlays/CollaboratorHintOverlayUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CollaboratorHintOverlayUtil_exports = {};
__export(CollaboratorHintOverlayUtil_exports, {
  CollaboratorHintOverlayUtil: () => CollaboratorHintOverlayUtil
});
module.exports = __toCommonJS(CollaboratorHintOverlayUtil_exports);
var import_editor = require("@tldraw/editor");
let _arrowPath = null;
function getArrowPath() {
  if (!_arrowPath) {
    _arrowPath = new Path2D("M -2 -5 L 2 0 L -2 5 Z");
  }
  return _arrowPath;
}
class CollaboratorHintOverlayUtil extends import_editor.OverlayUtil {
  static type = "collaborator_hint";
  options = { zIndex: 900, lineWidth: 3, viewportPadding: 5 };
  isActive() {
    const viewport = this.editor.getViewportPageBounds();
    const zoom = this.editor.getZoomLevel();
    return this.editor.getVisibleCollaboratorsOnCurrentPage().some((presence) => {
      const { cursor } = presence;
      if (!cursor) return false;
      return !this._isCursorInViewport(cursor, viewport, zoom);
    });
  }
  getOverlays() {
    const overlays = [];
    const viewport = this.editor.getViewportPageBounds();
    const zoom = this.editor.getZoomLevel();
    for (const presence of this.editor.getVisibleCollaboratorsOnCurrentPage()) {
      const { cursor, color, userId } = presence;
      if (!cursor) continue;
      if (this._isCursorInViewport(cursor, viewport, zoom)) continue;
      const pad = this.options.viewportPadding / zoom;
      const x = (0, import_editor.clamp)(cursor.x, viewport.minX + pad, viewport.maxX - pad);
      const y = (0, import_editor.clamp)(cursor.y, viewport.minY + pad, viewport.maxY - pad);
      const rotation = import_editor.Vec.Angle(viewport.center, cursor);
      overlays.push({
        id: `collaborator_hint:${userId}`,
        type: "collaborator_hint",
        props: { x, y, rotation, color }
      });
    }
    return overlays;
  }
  render(ctx, overlays) {
    const zoom = this.editor.getZoomLevel();
    const scale = 1 / zoom;
    for (const overlay of overlays) {
      const { x, y, rotation, color } = overlay.props;
      ctx.save();
      ctx.translate(x, y);
      ctx.scale(scale, scale);
      ctx.rotate(rotation);
      const path = getArrowPath();
      ctx.lineWidth = this.options.lineWidth;
      ctx.strokeStyle = "#ffffff";
      ctx.stroke(path);
      ctx.fillStyle = color;
      ctx.fill(path);
      ctx.restore();
    }
  }
  /** @internal */
  _isCursorInViewport(cursor, viewport, zoom) {
    return !(cursor.x < viewport.minX - 12 / zoom || cursor.y < viewport.minY - 16 / zoom || cursor.x > viewport.maxX - 12 / zoom || cursor.y > viewport.maxY - 16 / zoom);
  }
}
