// tldraw/dist-cjs/lib/overlays/CollaboratorCursorOverlayUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CollaboratorCursorOverlayUtil_exports = {};
__export(CollaboratorCursorOverlayUtil_exports, {
  CollaboratorCursorOverlayUtil: () => CollaboratorCursorOverlayUtil
});
module.exports = __toCommonJS(CollaboratorCursorOverlayUtil_exports);
var import_editor = require("@tldraw/editor");
let _shadowHead = null;
let _shadowTail = null;
let _whiteHead = null;
let _whiteTail = null;
let _fillHead = null;
let _fillTail = null;
function getCursorPaths() {
  if (!_shadowHead) {
    _shadowHead = new Path2D("M 1 13.4219 V -2.593 L 12.591 9.026 H 5.81 L 5.399 9.15 Z");
    _shadowTail = new Path2D("M 10.0595 13.9498 L 6.3715 15.4978 L 2.4965 6.2803 L 6.1845 4.7323 Z");
    _whiteHead = new Path2D("M 0 12.4219 V -3.593 L 11.591 8.026 H 4.81 L 4.399 8.15 Z");
    _whiteTail = new Path2D("M 9.0595 12.9498 L 5.3715 14.4978 L 1.4965 5.2803 L 5.1845 3.7323 Z");
    _fillHead = new Path2D("M 1 -1.186 V 10.002 L 3.969 7.136 L 4.397 6.997 H 9.165 Z");
    _fillTail = new Path2D("M 7.751 12.4155 L 5.907 13.1895 L 2.807 5.8155 L 4.648 5.0405 Z");
  }
  return {
    shadowHead: _shadowHead,
    shadowTail: _shadowTail,
    whiteHead: _whiteHead,
    whiteTail: _whiteTail,
    fillHead: _fillHead,
    fillTail: _fillTail
  };
}
const TRUNCATE_CACHE_MAX = 200;
const DEFAULT_LABEL_FONT_FAMILY = "'tldraw_sans', sans-serif";
function getLabelFontFamily(editorContainer, editorWindow) {
  const fontFamily = editorWindow.getComputedStyle(editorContainer).getPropertyValue("--tl-font-sans").trim();
  return fontFamily && !fontFamily.includes("var(") ? fontFamily : DEFAULT_LABEL_FONT_FAMILY;
}
class CollaboratorCursorOverlayUtil extends import_editor.OverlayUtil {
  static type = "collaborator_cursor";
  options = { zIndex: 1100, fontSize: 12, nameMaxWidth: 120, chatMaxWidth: 200 };
  // Cache truncated text results to avoid repeated measureText loops.
  // Key format: `${maxWidth}|${text}` with an upper bound on cache size.
  // Per-editor so multiple <Tldraw /> instances on one page don't trample
  // each other's entries.
  _truncateCache = /* @__PURE__ */ new Map();
  isActive() {
    return this.editor.getVisibleCollaboratorsOnCurrentPage().some((presence) => !!presence.cursor);
  }
  getOverlays() {
    const overlays = [];
    for (const presence of this.editor.getVisibleCollaboratorsOnCurrentPage()) {
      const { cursor, color, userName, chatMessage, userId } = presence;
      if (!cursor) continue;
      overlays.push({
        id: `collaborator_cursor:${userId}`,
        type: "collaborator_cursor",
        props: {
          x: cursor.x,
          y: cursor.y,
          color,
          name: userName !== "New User" ? userName : null,
          chatMessage: chatMessage ?? ""
        }
      });
    }
    return overlays;
  }
  render(ctx, overlays) {
    const zoom = this.editor.getZoomLevel();
    const scale = 1 / zoom;
    const viewport = this.editor.getViewportPageBounds();
    const viewportMarginX = 12 / zoom;
    const viewportMarginY = 16 / zoom;
    const labelFontFamily = getLabelFontFamily(
      this.editor.getContainer(),
      this.editor.getContainerWindow()
    );
    let paths = null;
    for (const overlay of overlays) {
      const { x, y, color, name, chatMessage } = overlay.props;
      if (x < viewport.minX - viewportMarginX || y < viewport.minY - viewportMarginY || x > viewport.maxX - viewportMarginX || y > viewport.maxY - viewportMarginY) {
        continue;
      }
      ctx.save();
      ctx.translate(x, y);
      ctx.scale(scale, scale);
      paths ??= getCursorPaths();
      ctx.fillStyle = "rgba(0,0,0,0.2)";
      ctx.fill(paths.shadowHead);
      ctx.fill(paths.shadowTail);
      ctx.fillStyle = "#ffffff";
      ctx.fill(paths.whiteHead);
      ctx.fill(paths.whiteTail);
      ctx.fillStyle = color;
      ctx.fill(paths.fillHead);
      ctx.fill(paths.fillTail);
      if (chatMessage) {
        if (name) {
          this._drawNameTitle(ctx, name, color, labelFontFamily);
        }
        this._drawChatBubble(ctx, chatMessage, color, labelFontFamily);
      } else if (name) {
        this._drawNameTag(ctx, name, color, labelFontFamily);
      }
      ctx.restore();
    }
  }
  /** Name tag (no chat) - colored background with white text */
  _drawNameTag(ctx, name, color, fontFamily) {
    const { fontSize, nameMaxWidth } = this.options;
    ctx.font = `${fontSize}px ${fontFamily}`;
    const text = this._truncateText(ctx, name, nameMaxWidth);
    const metrics = ctx.measureText(text);
    const textWidth = Math.min(metrics.width, nameMaxWidth);
    const px = 6;
    const py = 3;
    const x = 13;
    const y = 16;
    const h = fontSize + py * 2;
    const w = textWidth + px * 2;
    ctx.fillStyle = color;
    ctx.beginPath();
    ctx.roundRect(x, y, w, h, 4);
    ctx.fill();
    ctx.fillStyle = "#ffffff";
    ctx.textBaseline = "top";
    ctx.fillText(text, x + px, y + py);
  }
  /** Name title (when chat is present) - text with shadow, no background */
  _drawNameTitle(ctx, name, color, fontFamily) {
    const { fontSize, nameMaxWidth } = this.options;
    ctx.font = `${fontSize}px ${fontFamily}`;
    const text = this._truncateText(ctx, name, nameMaxWidth);
    const x = 13;
    const y = -2;
    ctx.strokeStyle = "#ffffff";
    ctx.lineWidth = 3;
    ctx.lineJoin = "round";
    ctx.textBaseline = "top";
    ctx.strokeText(text, x, y);
    ctx.fillStyle = color;
    ctx.fillText(text, x, y);
  }
  /** Chat bubble - colored background with white text */
  _drawChatBubble(ctx, chatMessage, color, fontFamily) {
    const { fontSize, chatMaxWidth } = this.options;
    ctx.font = `${fontSize}px ${fontFamily}`;
    const text = this._truncateText(ctx, chatMessage, chatMaxWidth);
    const metrics = ctx.measureText(text);
    const textWidth = Math.min(metrics.width, chatMaxWidth);
    const px = 6;
    const py = 3;
    const x = 13;
    const y = 16;
    const h = fontSize + py * 2;
    const w = textWidth + px * 2;
    ctx.fillStyle = color;
    ctx.beginPath();
    ctx.roundRect(x, y, w, h, 4);
    ctx.fill();
    ctx.fillStyle = "#ffffff";
    ctx.textBaseline = "top";
    ctx.fillText(text, x + px, y + py);
  }
  renderMinimap(ctx, overlays, zoom) {
    const radius = 3 / zoom;
    for (const overlay of overlays) {
      const { x, y, color } = overlay.props;
      ctx.beginPath();
      ctx.arc(x, y, radius, 0, import_editor.PI2);
      ctx.fillStyle = color;
      ctx.fill();
    }
  }
  _truncateText(ctx, text, maxWidth) {
    const key = `${ctx.font}|${maxWidth}|${text}`;
    const cached = this._truncateCache.get(key);
    if (cached !== void 0) return cached;
    if (ctx.measureText(text).width <= maxWidth) {
      return this._setTruncatedTextCache(key, text);
    }
    const ellipsis = "\u2026";
    let low = 0;
    let high = text.length;
    while (low < high) {
      const mid = Math.ceil((low + high) / 2);
      if (ctx.measureText(text.slice(0, mid) + ellipsis).width <= maxWidth) {
        low = mid;
      } else {
        high = mid - 1;
      }
    }
    return this._setTruncatedTextCache(key, text.slice(0, low) + ellipsis);
  }
  _setTruncatedTextCache(key, result) {
    if (this._truncateCache.size >= TRUNCATE_CACHE_MAX) this._truncateCache.clear();
    this._truncateCache.set(key, result);
    return result;
  }
}
