// tldraw/dist-cjs/lib/overlays/CollaboratorScribbleOverlayUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CollaboratorScribbleOverlayUtil_exports = {};
__export(CollaboratorScribbleOverlayUtil_exports, {
  CollaboratorScribbleOverlayUtil: () => CollaboratorScribbleOverlayUtil
});
module.exports = __toCommonJS(CollaboratorScribbleOverlayUtil_exports);
var import_editor = require("@tldraw/editor");
var import_getStroke = require("../shapes/shared/freehand/getStroke");
class CollaboratorScribbleOverlayUtil extends import_editor.OverlayUtil {
  static type = "collaborator_scribble";
  options = { zIndex: 800, streamline: 0.32, cacheSize: 500 };
  // String-keyed (not a WeakMap) because the cache key is a logical identity
  // — `${overlay.id}` derived from `scribble.id` — not the scribble object.
  // Tldraw's store replaces record objects on every update, so a WeakMap
  // keyed on the `TLScribble` instance would cache-miss every frame. Lifetime
  // is bounded by the Util instance (so, by the editor) plus the `cacheSize`
  // cap in `render` below.
  _collabScribblePathCache = /* @__PURE__ */ new Map();
  isActive() {
    return this.editor.getVisibleCollaboratorsOnCurrentPage().some((c) => c.scribbles.length > 0);
  }
  getOverlays() {
    const overlays = [];
    for (const presence of this.editor.getVisibleCollaboratorsOnCurrentPage()) {
      const { scribbles, color, userId } = presence;
      for (const scribble of scribbles) {
        overlays.push({
          id: `collaborator_scribble:${userId}:${scribble.id}`,
          type: "collaborator_scribble",
          props: { scribble, color }
        });
      }
    }
    return overlays;
  }
  render(ctx, overlays) {
    const zoom = this.editor.getZoomLevel();
    for (const overlay of overlays) {
      const { scribble, color } = overlay.props;
      const ptsLen = scribble.points.length;
      if (!ptsLen) continue;
      const last = scribble.points[ptsLen - 1];
      const cacheKey = overlay.id;
      const cached = this._collabScribblePathCache.get(cacheKey);
      let path;
      if (cached && cached.len === ptsLen && cached.lastX === last.x && cached.lastY === last.y && cached.zoom === zoom && cached.size === scribble.size && cached.taper === scribble.taper && cached.state === scribble.state) {
        path = cached.path;
      } else {
        const stroke = (0, import_getStroke.getStroke)(scribble.points, {
          size: scribble.size / zoom,
          start: { taper: scribble.taper, easing: import_editor.EASINGS.linear },
          last: scribble.state === "complete" || scribble.state === "stopping",
          simulatePressure: false,
          streamline: this.options.streamline
        });
        let d;
        if (stroke.length < 4) {
          const r = scribble.size / zoom / 2;
          const { x, y } = last;
          d = `M ${x - r},${y} a ${r},${r} 0 1,0 ${r * 2},0 a ${r},${r} 0 1,0 ${-r * 2},0`;
        } else {
          d = (0, import_editor.getSvgPathFromPoints)(stroke);
        }
        path = new Path2D(d);
        this._collabScribblePathCache.set(cacheKey, {
          len: ptsLen,
          lastX: last.x,
          lastY: last.y,
          zoom,
          size: scribble.size,
          taper: scribble.taper,
          state: scribble.state,
          path
        });
        if (this._collabScribblePathCache.size > this.options.cacheSize)
          this._collabScribblePathCache.clear();
      }
      ctx.fillStyle = color;
      ctx.globalAlpha = scribble.color === "laser" ? 0.5 : 0.1;
      ctx.fill(path);
      ctx.globalAlpha = 1;
    }
  }
}
