// tldraw/dist-cjs/lib/overlays/CollaboratorShapeIndicatorOverlayUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CollaboratorShapeIndicatorOverlayUtil_exports = {};
__export(CollaboratorShapeIndicatorOverlayUtil_exports, {
  CollaboratorShapeIndicatorOverlayUtil: () => CollaboratorShapeIndicatorOverlayUtil
});
module.exports = __toCommonJS(CollaboratorShapeIndicatorOverlayUtil_exports);
var import_editor = require("@tldraw/editor");
class CollaboratorShapeIndicatorOverlayUtil extends import_editor.OverlayUtil {
  static type = "collaborator_shape_indicator";
  options = { zIndex: 49, lineWidth: 1.5, alpha: 0.7 };
  isActive() {
    return this.editor.getVisibleCollaboratorsOnCurrentPage().some((presence) => presence.selectedShapeIds.length > 0);
  }
  getOverlays() {
    const editor = this.editor;
    const selectedPresences = editor.getVisibleCollaboratorsOnCurrentPage().filter((presence) => presence.selectedShapeIds.length > 0);
    if (selectedPresences.length === 0) return [];
    const renderingShapeIds = new Set(editor.getRenderingShapes().map((s) => s.id));
    const indicators = [];
    for (const presence of selectedPresences) {
      const visibleShapeIds = presence.selectedShapeIds.filter(
        (id) => renderingShapeIds.has(id) && !editor.isShapeHidden(id)
      );
      if (visibleShapeIds.length === 0) continue;
      indicators.push({ color: presence.color, shapeIds: visibleShapeIds });
    }
    if (indicators.length === 0) return [];
    return [
      {
        id: "collaborator_shape_indicator",
        type: "collaborator_shape_indicator",
        props: { indicators }
      }
    ];
  }
  render(ctx, overlays) {
    const overlay = overlays[0];
    if (!overlay) return;
    const editor = this.editor;
    const zoom = editor.getZoomLevel();
    ctx.lineCap = "round";
    ctx.lineJoin = "round";
    ctx.lineWidth = this.options.lineWidth / zoom;
    ctx.globalAlpha = this.options.alpha;
    for (const { color, shapeIds } of overlay.props.indicators) {
      ctx.strokeStyle = color;
      (0, import_editor.strokeShapeIndicators)(editor, ctx, shapeIds);
    }
    ctx.globalAlpha = 1;
  }
}
