// tldraw/dist-cjs/lib/overlays/CollaboratorBrushOverlayUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var CollaboratorBrushOverlayUtil_exports = {};
__export(CollaboratorBrushOverlayUtil_exports, {
  CollaboratorBrushOverlayUtil: () => CollaboratorBrushOverlayUtil
});
module.exports = __toCommonJS(CollaboratorBrushOverlayUtil_exports);
var import_editor = require("@tldraw/editor");
class CollaboratorBrushOverlayUtil extends import_editor.OverlayUtil {
  static type = "collaborator_brush";
  options = { zIndex: 700, lineWidth: 1 };
  isActive() {
    return this.editor.getVisibleCollaboratorsOnCurrentPage().some((c) => c.brush !== null);
  }
  getOverlays() {
    const overlays = [];
    for (const presence of this.editor.getVisibleCollaboratorsOnCurrentPage()) {
      const { brush, color, userId } = presence;
      if (!brush) continue;
      overlays.push({
        id: `collaborator_brush:${userId}`,
        type: "collaborator_brush",
        props: {
          x: brush.x,
          y: brush.y,
          w: Math.max(1, brush.w),
          h: Math.max(1, brush.h),
          color
        }
      });
    }
    return overlays;
  }
  render(ctx, overlays) {
    const zoom = this.editor.getZoomLevel();
    for (const overlay of overlays) {
      const { x, y, w, h, color } = overlay.props;
      ctx.globalAlpha = 0.075;
      ctx.fillStyle = color;
      ctx.fillRect(x, y, w, h);
      ctx.globalAlpha = 0.1;
      ctx.strokeStyle = color;
      ctx.lineWidth = this.options.lineWidth / zoom;
      ctx.strokeRect(x, y, w, h);
      ctx.globalAlpha = 1;
    }
  }
  renderMinimap(ctx, overlays) {
    for (const overlay of overlays) {
      const { x, y, w, h, color } = overlay.props;
      ctx.globalAlpha = 0.75;
      ctx.fillStyle = color;
      ctx.fillRect(x, y, w, h);
    }
    ctx.globalAlpha = 1;
  }
}
