// tldraw/dist-cjs/lib/overlays/ScribbleOverlayUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ScribbleOverlayUtil_exports = {};
__export(ScribbleOverlayUtil_exports, {
  ScribbleOverlayUtil: () => ScribbleOverlayUtil
});
module.exports = __toCommonJS(ScribbleOverlayUtil_exports);
var import_editor = require("@tldraw/editor");
var import_getStroke = require("../shapes/shared/freehand/getStroke");
class ScribbleOverlayUtil extends import_editor.OverlayUtil {
  static type = "scribble";
  options = { zIndex: 600, streamline: 0.32, cacheSize: 500 };
  // Per-editor cache so multiple <Tldraw /> instances on one page don't
  // trample each other's entries. String-keyed (not a WeakMap) because the
  // cache key is a logical identity — `scribble.id` — not the scribble
  // object. Tldraw's store replaces record objects on every update, so a
  // WeakMap keyed on the `TLScribble` instance would cache-miss every frame.
  // Lifetime is bounded by the Util instance plus the `cacheSize` cap below.
  _scribblePathCache = /* @__PURE__ */ new Map();
  isActive() {
    return this.editor.getInstanceState().scribbles.length > 0;
  }
  getOverlays() {
    return this.editor.getInstanceState().scribbles.map((scribble) => ({
      id: `scribble:${scribble.id}`,
      type: "scribble",
      props: { scribble }
    }));
  }
  render(ctx, overlays) {
    const zoom = this.editor.getZoomLevel();
    const colors = this.editor.getCurrentTheme().colors[this.editor.getColorMode()];
    for (const overlay of overlays) {
      const { scribble } = overlay.props;
      const ptsLen = scribble.points.length;
      if (!ptsLen) continue;
      const last = scribble.points[ptsLen - 1];
      const cached = this._scribblePathCache.get(scribble.id);
      let path;
      if (cached && cached.len === ptsLen && cached.lastX === last.x && cached.lastY === last.y && cached.zoom === zoom && cached.size === scribble.size && cached.taper === scribble.taper && cached.state === scribble.state) {
        path = cached.path;
      } else {
        const stroke = (0, import_getStroke.getStroke)(scribble.points, {
          size: scribble.size / zoom,
          start: { taper: scribble.taper, easing: import_editor.EASINGS.linear },
          last: scribble.state === "complete" || scribble.state === "stopping",
          simulatePressure: false,
          streamline: this.options.streamline
        });
        let d;
        if (stroke.length < 4) {
          const r = scribble.size / zoom / 2;
          const { x, y } = last;
          d = `M ${x - r},${y} a ${r},${r} 0 1,0 ${r * 2},0 a ${r},${r} 0 1,0 ${-r * 2},0`;
        } else {
          d = (0, import_editor.getSvgPathFromPoints)(stroke);
        }
        path = new Path2D(d);
        this._scribblePathCache.set(scribble.id, {
          len: ptsLen,
          lastX: last.x,
          lastY: last.y,
          zoom,
          size: scribble.size,
          taper: scribble.taper,
          state: scribble.state,
          path
        });
        if (this._scribblePathCache.size > this.options.cacheSize) this._scribblePathCache.clear();
      }
      ctx.fillStyle = resolveCanvasUiColor(colors, scribble.color);
      ctx.globalAlpha = scribble.opacity;
      ctx.fill(path);
      ctx.globalAlpha = 1;
    }
  }
}
function resolveCanvasUiColor(colors, color) {
  switch (color) {
    case "accent":
    case "selection-stroke":
      return colors.selectionStroke;
    case "selection-fill":
      return colors.selectionFill;
    case "white":
      return colors.selectedContrast;
    case "black":
      return colors.text;
    case "laser":
      return colors.laser;
    case "muted-1":
      return colors.brushFill;
    default:
      return colors.text;
  }
}
