// tldraw/dist-cjs/lib/overlays/ZoomBrushOverlayUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ZoomBrushOverlayUtil_exports = {};
__export(ZoomBrushOverlayUtil_exports, {
  ZoomBrushOverlayUtil: () => ZoomBrushOverlayUtil
});
module.exports = __toCommonJS(ZoomBrushOverlayUtil_exports);
var import_editor = require("@tldraw/editor");
class ZoomBrushOverlayUtil extends import_editor.OverlayUtil {
  static type = "zoom_brush";
  options = { zIndex: 400, lineWidth: 1 };
  isActive() {
    return this.editor.getInstanceState().zoomBrush !== null;
  }
  getOverlays() {
    const brush = this.editor.getInstanceState().zoomBrush;
    if (!brush) return [];
    return [
      {
        id: "zoom_brush",
        type: "zoom_brush",
        props: {
          x: brush.x,
          y: brush.y,
          w: Math.max(1, brush.w),
          h: Math.max(1, brush.h)
        }
      }
    ];
  }
  render(ctx, overlays) {
    const overlay = overlays[0];
    if (!overlay) return;
    const { x, y, w, h } = overlay.props;
    const zoom = this.editor.getZoomLevel();
    const colors = this.editor.getCurrentTheme().colors[this.editor.getColorMode()];
    ctx.fillStyle = colors.brushFill;
    ctx.fillRect(x, y, w, h);
    ctx.lineWidth = this.options.lineWidth / zoom;
    ctx.strokeStyle = colors.brushStroke;
    ctx.strokeRect(x, y, w, h);
  }
  renderMinimap(ctx, overlays, zoom) {
    const overlay = overlays[0];
    if (!overlay) return;
    const { x, y, w, h } = overlay.props;
    const colors = this.editor.getCurrentTheme().colors[this.editor.getColorMode()];
    ctx.fillStyle = colors.brushFill;
    ctx.fillRect(x, y, w, h);
    ctx.lineWidth = this.options.lineWidth / zoom;
    ctx.strokeStyle = colors.brushStroke;
    ctx.strokeRect(x, y, w, h);
  }
}
