// tldraw/dist-cjs/lib/shapes/geo/GeoShapeBody.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var GeoShapeBody_exports = {};
__export(GeoShapeBody_exports, {
  GeoShapeBody: () => GeoShapeBody
});
module.exports = __toCommonJS(GeoShapeBody_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_PatternFill = require("../shared/PatternFill");
var import_getGeoShapePath = require("./getGeoShapePath");
function GeoShapeBody({
  shape,
  shouldScale,
  forceSolid,
  strokeColor,
  strokeWidth: unscaledStrokeWidth,
  fillColor,
  patternFillFallbackColor,
  customGeoTypes
}) {
  const scaleToUse = shouldScale ? shape.props.scale : 1;
  const strokeWidth = unscaledStrokeWidth * scaleToUse;
  const { dash, fill } = shape.props;
  const path = (0, import_getGeoShapePath.getGeoShapePath)(shape, unscaledStrokeWidth, customGeoTypes);
  const fillPath = dash === "draw" && !forceSolid ? path.toDrawD({ strokeWidth, randomSeed: shape.id, passes: 1, offset: 0, onlyFilled: true }) : path.toD({ onlyFilled: true });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
    fill === "none" ? null : fill === "pattern" ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_PatternFill.PatternFill,
      {
        d: fillPath,
        fillColor,
        patternFillFallbackColor,
        scale: scaleToUse
      }
    ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)("path", { fill: fillColor, d: fillPath }),
    path.toSvg({
      style: dash,
      strokeWidth,
      forceSolid,
      randomSeed: shape.id,
      props: { fill: "none", stroke: strokeColor }
    })
  ] });
}
