// tldraw/dist-cjs/lib/tools/selection-logic/updateHoveredShapeId.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var updateHoveredShapeId_exports = {};
__export(updateHoveredShapeId_exports, {
  cancelUpdateHoveredShapeId: () => cancelUpdateHoveredShapeId,
  updateHoveredShapeId: () => updateHoveredShapeId
});
module.exports = __toCommonJS(updateHoveredShapeId_exports);
var import_editor = require("@tldraw/editor");
const hoverLockedEditors = /* @__PURE__ */ new WeakMap();
function getShapeToHover(editor) {
  const hitShape = editor.getShapeAtPoint(editor.inputs.getCurrentPagePoint(), {
    hitInside: false,
    hitLabels: false,
    hitLocked: editor.options.selectLockedShapes,
    margin: editor.options.hitTestMargin / editor.getZoomLevel(),
    renderingOnly: true
  });
  if (!hitShape) return null;
  let shapeToHover = void 0;
  const outermostShape = editor.getOutermostSelectableShape(hitShape);
  if (outermostShape === hitShape) {
    shapeToHover = hitShape;
  } else {
    if (outermostShape.id === editor.getFocusedGroupId() || editor.getSelectedShapeIds().includes(outermostShape.id)) {
      shapeToHover = hitShape;
    } else {
      shapeToHover = outermostShape;
    }
  }
  return shapeToHover.id;
}
function _updateHoveredShapeId(editor) {
  if (editor.isDisposed) return;
  const cameraMoving = editor.getCameraState() === "moving";
  if (!cameraMoving) {
    hoverLockedEditors.set(editor, false);
    const nextHoveredId2 = getShapeToHover(editor);
    return editor.setHoveredShape(nextHoveredId2);
  }
  if (hoverLockedEditors.get(editor)) {
    return void 0;
  }
  const currentHoveredId = editor.getHoveredShapeId();
  if (!currentHoveredId) {
    hoverLockedEditors.set(editor, true);
    return void 0;
  }
  const nextHoveredId = getShapeToHover(editor);
  if (nextHoveredId === currentHoveredId) {
    return void 0;
  }
  editor.setHoveredShape(null);
  hoverLockedEditors.set(editor, true);
  return void 0;
}
const THROTTLE_MS = process.env.NODE_ENV === "test" ? 0 : 32;
const editorThrottles = /* @__PURE__ */ new WeakMap();
function getThrottled(editor) {
  let throttled = editorThrottles.get(editor);
  if (!throttled) {
    throttled = (0, import_editor.throttle)(_updateHoveredShapeId, THROTTLE_MS);
    editorThrottles.set(editor, throttled);
  }
  return throttled;
}
function updateHoveredShapeId(editor) {
  getThrottled(editor)(editor);
}
function cancelUpdateHoveredShapeId(editor) {
  editorThrottles.get(editor)?.cancel();
}
