// tldraw/dist-cjs/lib/tools/HandTool/childStates/OneFingerZooming.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OneFingerZooming_exports = {};
__export(OneFingerZooming_exports, {
  OneFingerZooming: () => OneFingerZooming
});
module.exports = __toCommonJS(OneFingerZooming_exports);
var import_editor = require("@tldraw/editor");
class OneFingerZooming extends import_editor.StateNode {
  static id = "one_finger_zooming";
  anchorScreenPoint = new import_editor.Vec();
  initialCamera = new import_editor.Vec();
  initialZoom = 1;
  originScreenY = 0;
  onEnter(_info) {
    const camera = this.editor.getCamera();
    this.initialCamera = import_editor.Vec.From(camera);
    this.initialZoom = camera.z;
    this.anchorScreenPoint = this.editor.inputs.getOriginScreenPoint().clone();
    this.originScreenY = this.editor.inputs.getCurrentScreenPoint().y;
    this.editor.setCursor({ type: "grab", rotation: 0 });
  }
  onPointerMove(_info) {
    this.editor.menus.clearOpenMenus();
    const currentScreenY = this.editor.inputs.getCurrentScreenPoint().y;
    const dy = (this.originScreenY - currentScreenY) * -1;
    const zoomFactor = Math.pow(2, dy / 200);
    const newZoom = this.clampZoom(this.initialZoom * zoomFactor);
    const { x: cx, y: cy } = this.initialCamera;
    const { x: sx, y: sy } = this.anchorScreenPoint;
    this.editor.setCamera(
      new import_editor.Vec(
        cx + sx / newZoom - sx / this.initialZoom,
        cy + sy / newZoom - sy / this.initialZoom,
        newZoom
      ),
      { immediate: true }
    );
  }
  onPointerUp(_info) {
    this.complete();
  }
  onCancel() {
    this.parent.transition("idle");
  }
  onInterrupt() {
    this.parent.transition("idle");
  }
  complete() {
    const pointerVelocity = this.editor.inputs.getPointerVelocity();
    const velocityAtPointerUp = Math.min(pointerVelocity.len(), 2);
    if (velocityAtPointerUp > 0.1) {
      this.editor.slideCamera({
        speed: velocityAtPointerUp,
        direction: { x: 0, y: 0, z: Math.sign(pointerVelocity.y) * 0.01 }
      });
    }
    this.parent.transition("idle");
  }
  clampZoom(zoom) {
    const { zoomSteps } = this.editor.getCameraOptions();
    const baseZoom = this.editor.getBaseZoom();
    const zoomMin = zoomSteps[0] * baseZoom;
    const zoomMax = (0, import_editor.last)(zoomSteps) * baseZoom;
    return (0, import_editor.clamp)(zoom, zoomMin, zoomMax);
  }
}
