// @tldraw/editor/dist-cjs/lib/TldrawEditor.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TldrawEditor_exports = {};
__export(TldrawEditor_exports, {
  ErrorScreen: () => ErrorScreen,
  LoadingScreen: () => LoadingScreen,
  TL_CONTAINER_CLASS: () => TL_CONTAINER_CLASS,
  TldrawEditor: () => TldrawEditor,
  useOnMount: () => useOnMount
});
module.exports = __toCommonJS(TldrawEditor_exports);
var import_jsx_runtime = (
  // Store is ready to go, whether externally synced or not
  require("react/jsx-runtime")
);
var import_store = require("@tldraw/store");
var import_tlschema = require("@tldraw/tlschema");
var import_utils = require("@tldraw/utils");
var import_classnames = __toESM(require("classnames"), 1);
var import_react = __toESM(require("react"), 1);
var import_version = require("../version");
var import_DefaultErrorFallback = require("./components/default-components/DefaultErrorFallback");
var import_ErrorBoundary = require("./components/ErrorBoundary");
var import_createTLCurrentUser = require("./config/createTLCurrentUser");
var import_Editor = require("./editor/Editor");
var import_ThemeManager = require("./editor/managers/ThemeManager/ThemeManager");
var import_EditorComponentsContext = require("./hooks/EditorComponentsContext");
var import_useContainer = require("./hooks/useContainer");
var import_useCursor = require("./hooks/useCursor");
var import_useDarkMode = require("./hooks/useDarkMode");
var import_useEditor = require("./hooks/useEditor");
var import_useEditorComponents = require("./hooks/useEditorComponents");
var import_useEvent = require("./hooks/useEvent");
var import_useForceUpdate = require("./hooks/useForceUpdate");
var import_useIdentity = require("./hooks/useIdentity");
var import_useLocalStore = require("./hooks/useLocalStore");
var import_useRefState = require("./hooks/useRefState");
var import_useStateAttribute = require("./hooks/useStateAttribute");
var import_useZoomCss = require("./hooks/useZoomCss");
var import_LicenseProvider = require("./license/LicenseProvider");
var import_Watermark = require("./license/Watermark");
var import_dom = require("./utils/dom");
const EMPTY_SHAPE_UTILS_ARRAY = [];
const EMPTY_BINDING_UTILS_ARRAY = [];
const EMPTY_ASSET_UTILS_ARRAY = [];
const EMPTY_TOOLS_ARRAY = [];
const TL_CONTAINER_CLASS = "tl-container";
const TldrawEditor = (0, import_react.memo)(function TldrawEditor2({
  store,
  components,
  className,
  user: _user,
  options: _options,
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  textOptions: _textOptions,
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  deepLinks: _deepLinks,
  ...rest
}) {
  const resolvedThemes = (0, import_ThemeManager.resolveThemes)(rest.themes);
  (0, import_tlschema.registerColorsFromThemes)(resolvedThemes);
  (0, import_tlschema.registerFontsFromThemes)(resolvedThemes);
  const [container, setContainer] = (0, import_react.useState)(null);
  const user = (0, import_react.useMemo)(() => _user ?? (0, import_createTLCurrentUser.createTLCurrentUser)(), [_user]);
  const ErrorFallback = components?.ErrorFallback === void 0 ? import_DefaultErrorFallback.DefaultErrorFallback : components?.ErrorFallback;
  const mergedOptions = (0, import_react.useMemo)(() => {
    let result = _options;
    if (_textOptions) {
      result = { ...result, text: result?.text ?? _textOptions };
    }
    if (_deepLinks !== void 0) {
      result = { ...result, deepLinks: result?.deepLinks ?? _deepLinks };
    }
    return result;
  }, [_options, _textOptions, _deepLinks]);
  const withDefaults = {
    ...rest,
    shapeUtils: rest.shapeUtils ?? EMPTY_SHAPE_UTILS_ARRAY,
    bindingUtils: rest.bindingUtils ?? EMPTY_BINDING_UTILS_ARRAY,
    assetUtils: rest.assetUtils ?? EMPTY_ASSET_UTILS_ARRAY,
    tools: rest.tools ?? EMPTY_TOOLS_ARRAY,
    components,
    options: (0, import_useIdentity.useShallowObjectIdentity)(mergedOptions)
  };
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "div",
    {
      ref: setContainer,
      "data-tldraw": import_version.version,
      draggable: false,
      className: (0, import_classnames.default)(`${TL_CONTAINER_CLASS} tl-theme__light`, className),
      tabIndex: -1,
      role: "application",
      "aria-label": _options?.branding ?? "tldraw",
      children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        import_ErrorBoundary.OptionalErrorBoundary,
        {
          fallback: ErrorFallback,
          onError: (error) => (0, import_utils.annotateError)(error, { tags: { origin: "react.tldraw-before-app" } }),
          children: container && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_LicenseProvider.LicenseProvider, { licenseKey: rest.licenseKey, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_useContainer.ContainerProvider, { container, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_useEditorComponents.EditorComponentsProvider, { overrides: components, children: store ? store instanceof import_store.Store ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TldrawEditorWithReadyStore, { ...withDefaults, store, user }) : (
            // Store is a synced store, so handle syncing stages internally
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TldrawEditorWithLoadingStore, { ...withDefaults, store, user })
          ) : (
            // We have no store (it's undefined) so create one and possibly sync it
            /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TldrawEditorWithOwnStore, { ...withDefaults, store, user })
          ) }) }) })
        }
      )
    }
  );
});
function TldrawEditorWithOwnStore(props) {
  const {
    defaultName,
    snapshot,
    initialData,
    shapeUtils,
    bindingUtils,
    assetUtils,
    persistenceKey,
    sessionId,
    user,
    assets,
    users,
    migrations,
    themes
  } = props;
  const syncedStore = (0, import_useLocalStore.useLocalStore)({
    shapeUtils,
    bindingUtils,
    assetUtils,
    initialData,
    persistenceKey,
    sessionId,
    defaultName,
    snapshot,
    assets,
    users,
    migrations,
    themes
  });
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TldrawEditorWithLoadingStore, { ...props, store: syncedStore, user });
}
const TldrawEditorWithLoadingStore = (0, import_react.memo)(function TldrawEditorBeforeLoading({
  store,
  user,
  ...rest
}) {
  const container = (0, import_useContainer.useContainer)();
  (0, import_react.useLayoutEffect)(() => {
    if (user.userPreferences.get().colorScheme === "dark") {
      container.classList.remove("tl-theme__light");
      container.classList.add("tl-theme__dark");
    }
  }, [container, user]);
  const { LoadingScreen: LoadingScreen2 } = (0, import_EditorComponentsContext.useEditorComponents)();
  switch (store.status) {
    case "error": {
      throw store.error;
    }
    case "loading": {
      return LoadingScreen2 ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(LoadingScreen2, {}) : null;
    }
    case "not-synced": {
      break;
    }
    case "synced-local": {
      break;
    }
    case "synced-remote": {
      break;
    }
  }
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(TldrawEditorWithReadyStore, { ...rest, store: store.store, user });
});
const noAutoFocus = () => (0, import_dom.getGlobalDocument)().location.search.includes("tldraw_preserve_focus");
function TldrawEditorWithReadyStore({
  onMount,
  children,
  store,
  tools,
  shapeUtils,
  bindingUtils,
  assetUtils,
  overlayUtils,
  user,
  initialState,
  autoFocus = true,
  // eslint-disable-next-line @typescript-eslint/no-deprecated
  cameraOptions,
  options,
  licenseKey,
  getShapeVisibility,
  colorScheme,
  assetUrls,
  themes,
  initialTheme
}) {
  const { ErrorFallback } = (0, import_EditorComponentsContext.useEditorComponents)();
  const container = (0, import_useContainer.useContainer)();
  const [editor, setEditor] = (0, import_useRefState.useRefState)(null);
  const canvasRef = (0, import_react.useRef)(null);
  const _deepLinks = options?.deepLinks;
  const deepLinks = (0, import_useIdentity.useShallowObjectIdentity)(_deepLinks === true ? {} : _deepLinks);
  const editorOptionsRef = (0, import_react.useRef)({
    // for these, it's because they're only used when the editor first mounts:
    autoFocus: autoFocus && !noAutoFocus(),
    initialState,
    colorScheme,
    // for these, it's because we keep them up to date in a separate effect:
    cameraOptions,
    deepLinks,
    themes,
    initialTheme
  });
  (0, import_react.useLayoutEffect)(() => {
    editorOptionsRef.current = {
      autoFocus: autoFocus && !noAutoFocus(),
      initialState,
      colorScheme,
      cameraOptions,
      deepLinks,
      themes,
      initialTheme
    };
  }, [autoFocus, initialState, colorScheme, cameraOptions, deepLinks, themes, initialTheme]);
  (0, import_react.useLayoutEffect)(
    () => {
      const {
        autoFocus: autoFocus2,
        initialState: initialState2,
        colorScheme: initColorScheme,
        cameraOptions: cameraOptions2,
        deepLinks: deepLinks2,
        themes: themes2,
        initialTheme: initialTheme2
      } = editorOptionsRef.current;
      const editor2 = new import_Editor.Editor({
        store,
        shapeUtils,
        bindingUtils,
        assetUtils,
        overlayUtils,
        tools,
        getContainer: () => container,
        user,
        initialState: initialState2,
        // we should check for some kind of query parameter that turns off autofocus
        autoFocus: autoFocus2,
        cameraOptions: cameraOptions2,
        options,
        licenseKey,
        getShapeVisibility,
        colorScheme: initColorScheme,
        fontAssetUrls: assetUrls?.fonts,
        themes: themes2,
        initialTheme: initialTheme2
      });
      editor2.updateViewportScreenBounds(canvasRef.current ?? container);
      if (deepLinks2) {
        if (!deepLinks2?.getUrl) {
          editor2.navigateToDeepLink(deepLinks2);
        } else {
          editor2.navigateToDeepLink({ ...deepLinks2, url: deepLinks2.getUrl(editor2) });
        }
      }
      setEditor(editor2);
      return () => {
        editor2.dispose();
      };
    },
    // if any of these change, we need to recreate the editor.
    [
      assetUtils,
      bindingUtils,
      colorScheme,
      container,
      options,
      overlayUtils,
      shapeUtils,
      store,
      tools,
      user,
      setEditor,
      licenseKey,
      getShapeVisibility,
      assetUrls
    ]
  );
  (0, import_react.useLayoutEffect)(() => {
    if (!editor) return;
    if (deepLinks) {
      return editor.registerDeepLinkListener(deepLinks);
    }
  }, [editor, deepLinks]);
  (0, import_react.useLayoutEffect)(() => {
    if (editor && (cameraOptions || options?.camera)) {
      editor.setCameraOptions({ ...cameraOptions, ...options?.camera });
    }
  }, [editor, cameraOptions, options?.camera]);
  (0, import_react.useLayoutEffect)(() => {
    if (editor && themes) {
      for (const def of Object.values(themes)) {
        editor.updateTheme(def);
      }
    }
  }, [editor, themes]);
  (0, import_react.useLayoutEffect)(() => {
    if (editor && initialTheme) {
      editor.setCurrentTheme(initialTheme);
    }
  }, [editor, initialTheme]);
  const crashingError = (0, import_react.useSyncExternalStore)(
    (0, import_react.useCallback)(
      (onStoreChange) => {
        if (editor) {
          editor.on("crash", onStoreChange);
          return () => editor.off("crash", onStoreChange);
        }
        return () => {
        };
      },
      [editor]
    ),
    () => editor?.getCrashingError() ?? null
  );
  (0, import_react.useEffect)(
    function handleFocusOnPointerDownForPreserveFocusMode() {
      if (!editor) return;
      function handleFocusOnPointerDown() {
        if (!editor) return;
        editor.focus();
      }
      function handleBlurOnPointerDown() {
        if (!editor) return;
        editor.blur();
      }
      if (autoFocus && noAutoFocus()) {
        const container2 = editor.getContainer();
        container2.addEventListener("pointerdown", handleFocusOnPointerDown);
        container2.ownerDocument.body.addEventListener("pointerdown", handleBlurOnPointerDown);
        return () => {
          container2.removeEventListener("pointerdown", handleFocusOnPointerDown);
          container2.ownerDocument.body.removeEventListener("pointerdown", handleBlurOnPointerDown);
        };
      }
    },
    [editor, autoFocus]
  );
  const [_fontLoadingState, setFontLoadingState] = (0, import_react.useState)(null);
  let fontLoadingState = _fontLoadingState;
  if (editor !== fontLoadingState?.editor) {
    fontLoadingState = null;
  }
  (0, import_react.useLayoutEffect)(() => {
    if (!editor) return;
    if (editor.options.maxFontsToLoadBeforeRender === 0) {
      setFontLoadingState({ editor, isLoaded: true });
      return;
    }
    let isCancelled = false;
    setFontLoadingState({ editor, isLoaded: false });
    editor.fonts.loadRequiredFontsForCurrentPage(editor.options.maxFontsToLoadBeforeRender).finally(() => {
      if (isCancelled) return;
      setFontLoadingState({ editor, isLoaded: true });
    });
    return () => {
      isCancelled = true;
    };
  }, [editor]);
  const { Canvas, LoadingScreen: LoadingScreen2 } = (0, import_EditorComponentsContext.useEditorComponents)();
  if (!editor || !fontLoadingState?.isLoaded) {
    return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_jsx_runtime.Fragment, { children: [
      LoadingScreen2 && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(LoadingScreen2, {}),
      /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "tl-canvas", ref: canvasRef })
    ] });
  }
  return (
    // the top-level tldraw component also renders an error boundary almost
    // identical to this one. the reason we have two is because this one has
    // access to `App`, which means that here we can enrich errors with data
    // from app for reporting, and also still attempt to render the user's
    // document in the event of an error to reassure them that their work is
    // not lost.
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_ErrorBoundary.OptionalErrorBoundary,
      {
        fallback: ErrorFallback,
        onError: (error) => editor.annotateError(error, { origin: "react.tldraw", willCrashApp: true }),
        children: crashingError ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Crash, { crashingError }) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_useEditor.EditorProvider, { editor, children: /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(Layout, { onMount, children: [
          children ?? (Canvas ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Canvas, {}, editor.contextId) : null),
          /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_Watermark.Watermark, {})
        ] }) })
      }
    )
  );
}
function Layout({ children, onMount }) {
  (0, import_useZoomCss.useZoomCss)();
  (0, import_useCursor.useCursor)();
  (0, import_useDarkMode.useDarkMode)();
  (0, import_useForceUpdate.useForceUpdate)();
  (0, import_useStateAttribute.useStateAttribute)();
  useOnMount((editor) => {
    const teardownStore = editor.store.props.onMount(editor);
    const teardownCallback = onMount?.(editor);
    return () => {
      teardownStore?.();
      teardownCallback?.();
    };
  });
  return children;
}
function Crash({ crashingError }) {
  throw crashingError;
}
function LoadingScreen({ children }) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "tl-loading", "aria-busy": "true", tabIndex: 0, children });
}
function ErrorScreen({ children }) {
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "tl-loading", children });
}
function useOnMount(onMount) {
  const editor = (0, import_useEditor.useEditor)();
  const onMountEvent = (0, import_useEvent.useEvent)((editor2) => {
    let teardown = void 0;
    editor2.run(
      () => {
        teardown = onMount?.(editor2);
        editor2.emit("mount");
      },
      { history: "ignore" }
    );
    window.tldrawReady = true;
    return teardown;
  });
  import_react.default.useLayoutEffect(() => {
    if (editor) return onMountEvent?.(editor);
  }, [editor, onMountEvent]);
}
