// tldraw/dist-cjs/lib/shapes/frame/components/FrameLabelInput.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var FrameLabelInput_exports = {};
__export(FrameLabelInput_exports, {
  FrameLabelInput: () => FrameLabelInput
});
module.exports = __toCommonJS(FrameLabelInput_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_editor = require("@tldraw/editor");
var import_react = require("react");
var import_constants = require("../../../ui/constants");
var import_breakpoints = require("../../../ui/context/breakpoints");
var import_useTranslation = require("../../../ui/hooks/useTranslation/useTranslation");
var import_frameHelpers = require("../frameHelpers");
const FrameLabelInput = (0, import_react.forwardRef)(({ id, name, isEditing }, ref) => {
  const editor = (0, import_editor.useEditor)();
  const breakpoint = (0, import_breakpoints.useBreakpoint)();
  const isCoarsePointer = (0, import_editor.useValue)(
    "isCoarsePointer",
    () => editor.getInstanceState().isCoarsePointer,
    [editor]
  );
  const shouldUseWindowPrompt = breakpoint < import_constants.PORTRAIT_BREAKPOINT.TABLET_SM && isCoarsePointer;
  const promptOpen = (0, import_react.useRef)(false);
  const msg = (0, import_useTranslation.useTranslation)();
  const handlePointerDown = (0, import_react.useCallback)(
    (e) => {
      if (isEditing) editor.markEventAsHandled(e);
    },
    [editor, isEditing]
  );
  const handleKeyDown = (0, import_react.useCallback)(
    (e) => {
      if (e.key === "Enter" && !e.nativeEvent.isComposing) {
        editor.markEventAsHandled(e);
        e.currentTarget.blur();
        editor.setEditingShape(null);
      }
    },
    [editor]
  );
  const renameFrame = (0, import_react.useCallback)(
    (value) => {
      const shape = editor.getShape(id);
      if (!shape) return;
      const name2 = shape.props.name;
      if (name2 === value) return;
      editor.updateShapes([
        {
          id,
          type: "frame",
          props: { name: value }
        }
      ]);
    },
    [id, editor]
  );
  const handleBlur = (0, import_react.useCallback)(
    (e) => {
      renameFrame(e.currentTarget.value);
    },
    [renameFrame]
  );
  const handleChange = (0, import_react.useCallback)(
    (e) => {
      renameFrame(e.currentTarget.value);
    },
    [renameFrame]
  );
  (0, import_react.useEffect)(() => {
    if (!isEditing) {
      promptOpen.current = false;
      return;
    }
    if (isEditing && shouldUseWindowPrompt && !promptOpen.current) {
      promptOpen.current = true;
      const shape = editor.getShape(id);
      const currentName = shape?.props.name ?? "";
      const newName = window.prompt(msg("action.rename"), currentName);
      promptOpen.current = false;
      if (newName !== null) renameFrame(newName);
      editor.setEditingShape(null);
    }
  }, [isEditing, shouldUseWindowPrompt, id, msg, renameFrame, editor]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(
    "div",
    {
      className: `tl-frame-label ${isEditing && !shouldUseWindowPrompt ? "tl-frame-label__editing" : ""}`,
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
          "input",
          {
            className: "tl-frame-name-input",
            ref,
            disabled: !isEditing || shouldUseWindowPrompt,
            readOnly: !isEditing || shouldUseWindowPrompt,
            style: { display: isEditing ? void 0 : "none" },
            value: name,
            autoFocus: !shouldUseWindowPrompt,
            onKeyDown: handleKeyDown,
            onBlur: handleBlur,
            onChange: handleChange,
            onPointerDown: handlePointerDown,
            draggable: false
          }
        ),
        (0, import_frameHelpers.defaultEmptyAs)(name, "Frame") + String.fromCharCode(8203)
      ]
    }
  );
});
