// tldraw/dist-cjs/lib/ui/components/Minimap/DefaultMinimap.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DefaultMinimap_exports = {};
__export(DefaultMinimap_exports, {
  DefaultMinimap: () => DefaultMinimap
});
module.exports = __toCommonJS(DefaultMinimap_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_editor = require("@tldraw/editor");
var React = __toESM(require("react"), 1);
var import_useTranslation = require("../../hooks/useTranslation/useTranslation");
var import_MinimapManager = require("./MinimapManager");
const CLICK_JITTER_THRESHOLD_SQ = 4;
function DefaultMinimap() {
  const editor = (0, import_editor.useEditor)();
  const container = (0, import_editor.useContainer)();
  const msg = (0, import_useTranslation.useTranslation)();
  const rCanvas = React.useRef(null);
  const rPointing = React.useRef(false);
  const rActivePointerId = React.useRef(null);
  const rOriginScreenPoint = React.useRef(new import_editor.Vec());
  const minimapRef = React.useRef(void 0);
  React.useEffect(() => {
    try {
      const minimap = new import_MinimapManager.MinimapManager(editor, rCanvas.current, container);
      minimapRef.current = minimap;
      return minimapRef.current.close;
    } catch (e) {
      editor.annotateError(e, {
        origin: "minimap",
        willCrashApp: false
      });
      editor.timers.setTimeout(() => {
        throw e;
      });
      return void 0;
    }
  }, [editor, container]);
  const onDoubleClick = React.useCallback(
    (e) => {
      if (!editor.getCurrentPageShapeIds().size) return;
      const minimap = minimapRef.current;
      if (!minimap) return;
      const { clientX: x, clientY: y } = e;
      const clampedPoint = minimap.minimapScreenPointToPagePoint(x, y, false, true);
      minimap.originPagePoint.setTo(clampedPoint);
      minimap.originPageCenter.setTo(editor.getViewportPageBounds().center);
      const point = minimap.minimapScreenPointToPagePoint(x, y, false, false);
      editor.centerOnPoint(point, { animation: { duration: editor.options.animationMediumMs } });
    },
    [editor]
  );
  const onPointerDown = React.useCallback(
    (e) => {
      const minimap = minimapRef.current;
      if (!minimap) return;
      if (e.button !== 0) return;
      const elm = e.currentTarget;
      (0, import_editor.setPointerCapture)(elm, e);
      if (!editor.getCurrentPageShapeIds().size) return;
      const { clientX: x, clientY: y } = e;
      rPointing.current = true;
      rActivePointerId.current = e.pointerId;
      rOriginScreenPoint.current = new import_editor.Vec(x, y);
      minimap.isInViewport = false;
      const point = minimap.minimapScreenPointToPagePoint(x, y, false, false);
      const vpPageBounds = editor.getViewportPageBounds();
      const commonBounds = minimap.getContentPageBounds();
      const allowedBounds = new import_editor.Box(
        commonBounds.x - vpPageBounds.width / 2,
        commonBounds.y - vpPageBounds.height / 2,
        commonBounds.width + vpPageBounds.width,
        commonBounds.height + vpPageBounds.height
      );
      if (allowedBounds.containsPoint(point) && !vpPageBounds.containsPoint(point)) {
        const delta = import_editor.Vec.Sub(vpPageBounds.center, vpPageBounds.point);
        const pagePoint = import_editor.Vec.Add(point, delta);
        minimap.originPagePoint.setTo(pagePoint);
        minimap.originPageCenter.setTo(point);
        editor.centerOnPoint(point, { animation: { duration: editor.options.animationMediumMs } });
      } else {
        const clampedPoint = minimap.minimapScreenPointToPagePoint(x, y, false, true);
        minimap.isInViewport = vpPageBounds.containsPoint(clampedPoint);
        minimap.originPagePoint.setTo(clampedPoint);
        minimap.originPageCenter.setTo(vpPageBounds.center);
      }
      const body = editor.getContainerDocument().body;
      function endDrag() {
        if (rActivePointerId.current !== null && elm.hasPointerCapture(rActivePointerId.current)) {
          elm.releasePointerCapture(rActivePointerId.current);
        }
        rPointing.current = false;
        rActivePointerId.current = null;
        body.removeEventListener("pointerup", endDrag);
        body.removeEventListener("pointercancel", endDrag);
        body.removeEventListener("contextmenu", endDrag, true);
      }
      body.addEventListener("pointerup", endDrag);
      body.addEventListener("pointercancel", endDrag);
      body.addEventListener("contextmenu", endDrag, true);
    },
    [editor]
  );
  const onPointerMove = React.useCallback(
    (e) => {
      const minimap = minimapRef.current;
      if (!minimap) return;
      const { clientX: x, clientY: y } = e;
      const point = minimap.minimapScreenPointToPagePoint(x, y, e.shiftKey, true);
      if (rPointing.current) {
        if (import_editor.Vec.Dist2(rOriginScreenPoint.current, new import_editor.Vec(x, y)) <= CLICK_JITTER_THRESHOLD_SQ) {
          return;
        }
        if (minimap.isInViewport) {
          const delta = minimap.originPagePoint.clone().sub(minimap.originPageCenter);
          editor.centerOnPoint(import_editor.Vec.Sub(point, delta));
          return;
        }
        editor.centerOnPoint(point);
      }
      const pagePoint = minimap.getMinimapPagePoint(x, y);
      const screenPoint = editor.pageToScreen(pagePoint);
      const info = {
        type: "pointer",
        target: "canvas",
        name: "pointer_move",
        ...(0, import_editor.getPointerInfo)(editor, e),
        point: screenPoint,
        isPen: editor.getInstanceState().isPenMode
      };
      editor.dispatch(info);
    },
    [editor]
  );
  const onWheel = React.useCallback(
    (e) => {
      const offset = (0, import_editor.normalizeWheel)(e);
      editor.dispatch({
        type: "wheel",
        name: "wheel",
        delta: offset,
        point: new import_editor.Vec(e.clientX, e.clientY),
        shiftKey: e.shiftKey,
        altKey: e.altKey,
        ctrlKey: e.metaKey || e.ctrlKey,
        metaKey: e.metaKey,
        accelKey: (0, import_editor.isAccelKey)(e)
      });
    },
    [editor]
  );
  const colorMode = (0, import_editor.useColorMode)();
  const currentThemeId = (0, import_editor.useValue)("current theme id", () => editor.getCurrentThemeId(), [editor]);
  React.useEffect(() => {
    editor.timers.setTimeout(() => {
      minimapRef.current?.updateColors();
      minimapRef.current?.render();
    });
  }, [colorMode, currentThemeId, editor]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)("div", { className: "tlui-minimap", children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "canvas",
    {
      ref: rCanvas,
      role: "img",
      "aria-label": msg("navigation-zone.minimap"),
      "data-testid": "minimap.canvas",
      className: "tlui-minimap__canvas",
      onDoubleClick,
      onPointerMove,
      onPointerDown,
      onWheelCapture: onWheel
    }
  ) });
}
