// tldraw/dist-cjs/lib/shapes/embed/EmbedShapeUtil.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var EmbedShapeUtil_exports = {};
__export(EmbedShapeUtil_exports, {
  EmbedShapeUtil: () => EmbedShapeUtil
});
module.exports = __toCommonJS(EmbedShapeUtil_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_editor = require("@tldraw/editor");
var import_defaultEmbedDefinitions = require("../../defaultEmbedDefinitions");
var import_embeds = require("../../utils/embeds/embeds");
var import_BookmarkShapeUtil = require("../bookmark/BookmarkShapeUtil");
var import_getDisplayValues = require("../shared/getDisplayValues");
var import_rotated_box_shadow = require("../shared/rotated-box-shadow");
const getSandboxPermissions = (permissions) => {
  return Object.entries(permissions).filter(([_perm, isEnabled]) => isEnabled).map(([perm]) => perm).join(" ");
};
class EmbedShapeUtil extends import_editor.BaseBoxShapeUtil {
  static type = "embed";
  static props = import_editor.embedShapeProps;
  static migrations = import_editor.embedShapeMigrations;
  options = {
    embedDefinitions: import_defaultEmbedDefinitions.DEFAULT_EMBED_DEFINITIONS,
    embedConfig: {},
    getDefaultDisplayValues() {
      return {
        showShadow: true
      };
    },
    getCustomDisplayValues() {
      return {};
    }
  };
  canEditWhileLocked(shape) {
    const result = this.getEmbedDefinition(shape.props.url);
    if (!result) return true;
    return result.definition.canEditWhileLocked ?? true;
  }
  static legacyEmbedDefinitions = null;
  /** @deprecated - Use `EmbedShapeUtil.configure({ embedDefinitions: [...] })` instead. */
  static setEmbedDefinitions(embedDefinitions) {
    EmbedShapeUtil.legacyEmbedDefinitions = embedDefinitions;
  }
  getEmbedDefs() {
    return EmbedShapeUtil.legacyEmbedDefinitions ?? this.options.embedDefinitions;
  }
  getEmbedDefinitions() {
    return this.getEmbedDefs();
  }
  getEmbedDefinition(url) {
    return (0, import_embeds.getEmbedInfo)(this.getEmbedDefs(), url, this.options.embedConfig);
  }
  getText(shape) {
    return shape.props.url;
  }
  getAriaDescriptor(shape) {
    const embedInfo = this.getEmbedDefinition(shape.props.url);
    return embedInfo?.definition.title;
  }
  hideSelectionBoundsFg(shape) {
    return !this.canResize(shape);
  }
  canEdit(shape) {
    return true;
  }
  canResize(shape) {
    return this.getEmbedDefinition(shape.props.url)?.definition?.doesResize ?? true;
  }
  canEditInReadonly(shape) {
    return true;
  }
  getDefaultProps() {
    return {
      w: 300,
      h: 300,
      url: ""
    };
  }
  getGeometry(shape) {
    return super.getGeometry(shape);
  }
  isAspectRatioLocked(shape) {
    const embedInfo = this.getEmbedDefinition(shape.props.url);
    return embedInfo?.definition.isAspectRatioLocked ?? false;
  }
  onResize(shape, info) {
    const isAspectRatioLocked = this.isAspectRatioLocked(shape);
    const embedInfo = this.getEmbedDefinition(shape.props.url);
    let minWidth = embedInfo?.definition.minWidth ?? 200;
    let minHeight = embedInfo?.definition.minHeight ?? 200;
    if (isAspectRatioLocked) {
      const aspectRatio = shape.props.w / shape.props.h;
      if (aspectRatio > 1) {
        minWidth *= aspectRatio;
      } else {
        minHeight /= aspectRatio;
      }
    }
    return (0, import_editor.resizeBox)(shape, info, { minWidth, minHeight });
  }
  component(shape) {
    const svgExport = (0, import_editor.useSvgExportContext)();
    const { w, h, url } = shape.props;
    const isEditing = (0, import_editor.useIsEditing)(shape.id);
    const colorMode = (0, import_editor.useColorMode)();
    const dv = (0, import_getDisplayValues.getDisplayValues)(this, shape, colorMode);
    const embedInfo = this.getEmbedDefinition(url);
    const isHoveringWhileEditingSameShape = (0, import_editor.useValue)(
      "is hovering",
      () => {
        const { editingShapeId, hoveredShapeId } = this.editor.getCurrentPageState();
        if (editingShapeId && hoveredShapeId !== editingShapeId) {
          const editingShape = this.editor.getShape(editingShapeId);
          if (editingShape && this.editor.isShapeOfType(editingShape, "embed")) {
            return true;
          }
        }
        return false;
      },
      []
    );
    const pageRotation = this.editor.getShapePageTransform(shape).rotation();
    if (svgExport) {
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_editor.HTMLContainer, { className: "tl-embed-container", id: shape.id, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        "div",
        {
          className: "tl-embed",
          style: {
            border: 0,
            boxShadow: dv.showShadow ? (0, import_rotated_box_shadow.getRotatedBoxShadow)(pageRotation) : "none",
            borderRadius: embedInfo?.definition.overrideOutlineRadius ?? 8,
            background: embedInfo?.definition.backgroundColor ?? "var(--tl-color-background)",
            width: w,
            height: h
          }
        }
      ) });
    }
    const isInteractive = isEditing || isHoveringWhileEditingSameShape;
    const isIframe = typeof window !== "undefined" && (window !== window.top || window.self !== window.parent);
    if (isIframe && embedInfo?.definition.type === "tldraw") return null;
    const sandbox = getSandboxPermissions({
      ...import_defaultEmbedDefinitions.embedShapePermissionDefaults,
      ...embedInfo ? embedInfo.definition.overridePermissions ?? {} : import_defaultEmbedDefinitions.unknownEmbedShapePermissionOverrides
    });
    if (embedInfo?.definition.type === "github_gist") {
      const idFromGistUrl = embedInfo.url.split("/").pop();
      if (!idFromGistUrl) throw Error("No gist id!");
      const gistSandbox = getSandboxPermissions({
        ...import_defaultEmbedDefinitions.embedShapePermissionDefaults,
        ...embedInfo?.definition?.overridePermissions ?? {},
        "allow-same-origin": false
      });
      return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_editor.HTMLContainer, { className: "tl-embed-container", id: shape.id, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
        Gist,
        {
          id: idFromGistUrl,
          sandbox: gistSandbox,
          width: (0, import_editor.toDomPrecision)(w),
          height: (0, import_editor.toDomPrecision)(h),
          isInteractive,
          pageRotation,
          showShadow: dv.showShadow
        }
      ) });
    }
    const iframeSrc = embedInfo?.embedUrl ?? url;
    return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_editor.HTMLContainer, { className: "tl-embed-container", id: shape.id, children: iframeSrc ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      "iframe",
      {
        className: "tl-embed",
        sandbox,
        src: iframeSrc,
        width: (0, import_editor.toDomPrecision)(w),
        height: (0, import_editor.toDomPrecision)(h),
        draggable: false,
        frameBorder: "0",
        referrerPolicy: "strict-origin-when-cross-origin",
        tabIndex: isEditing ? 0 : -1,
        allowFullScreen: true,
        style: {
          border: 0,
          pointerEvents: isInteractive ? "auto" : "none",
          // Fix for safari <https://stackoverflow.com/a/49150908>
          zIndex: isInteractive ? "" : "-1",
          boxShadow: dv.showShadow ? (0, import_rotated_box_shadow.getRotatedBoxShadow)(pageRotation) : "none",
          borderRadius: embedInfo?.definition?.overrideOutlineRadius ?? 8,
          background: embedInfo?.definition?.backgroundColor
        }
      }
    ) : /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_BookmarkShapeUtil.BookmarkShapeComponent,
      {
        url,
        h,
        rotation: pageRotation,
        assetId: null,
        showImageContainer: false
      }
    ) });
  }
  getIndicatorPath(shape) {
    const path = new Path2D();
    path.rect(0, 0, shape.props.w, shape.props.h);
    return path;
  }
  getInterpolatedProps(startShape, endShape, t) {
    return {
      ...t > 0.5 ? endShape.props : startShape.props,
      w: (0, import_editor.lerp)(startShape.props.w, endShape.props.w, t),
      h: (0, import_editor.lerp)(startShape.props.h, endShape.props.h, t)
    };
  }
}
function Gist({
  id,
  sandbox,
  isInteractive,
  width,
  height,
  style,
  pageRotation,
  showShadow
}) {
  if (!id.match(/^[0-9a-f]+$/)) throw Error("No gist id!");
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
    "iframe",
    {
      className: "tl-embed",
      sandbox,
      draggable: false,
      width: (0, import_editor.toDomPrecision)(width),
      height: (0, import_editor.toDomPrecision)(height),
      frameBorder: "0",
      scrolling: "no",
      referrerPolicy: "strict-origin-when-cross-origin",
      tabIndex: isInteractive ? 0 : -1,
      style: {
        ...style,
        pointerEvents: isInteractive ? "all" : "none",
        // Fix for safari <https://stackoverflow.com/a/49150908>
        zIndex: isInteractive ? "" : "-1",
        boxShadow: showShadow ? (0, import_rotated_box_shadow.getRotatedBoxShadow)(pageRotation) : "none"
      },
      srcDoc: `
			<html>
				<head>
					<base target="_blank">
				</head>
				<body>
					<script src=${`https://gist.github.com/${id}.js`}></script>
					<style type="text/css">
						* { margin: 0px; }
						table { height: 100%; background-color: red; }
						.gist { background-color: none; height: 100%;  }
						.gist .gist-file { height: calc(100vh - 2px); padding: 0px; display: grid; grid-template-rows: 1fr auto; }
					</style>
				</body>
			</html>`
    }
  );
}
