// tldraw/dist-cjs/lib/ui/components/ContextMenu/DefaultContextMenu.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var DefaultContextMenu_exports = {};
__export(DefaultContextMenu_exports, {
  DefaultContextMenu: () => DefaultContextMenu
});
module.exports = __toCommonJS(DefaultContextMenu_exports);
var import_jsx_runtime = require("react/jsx-runtime");
var import_editor = require("@tldraw/editor");
var import_radix_ui = require("radix-ui");
var import_react = require("react");
var import_useMenuIsOpen = require("../../hooks/useMenuIsOpen");
var import_useTranslation = require("../../hooks/useTranslation/useTranslation");
var import_TldrawUiMenuContext = require("../primitives/menus/TldrawUiMenuContext");
var import_DefaultContextMenuContent = require("./DefaultContextMenuContent");
const DefaultContextMenu = (0, import_react.memo)(function DefaultContextMenu2({
  children,
  disabled = false
}) {
  const editor = (0, import_editor.useEditor)();
  const msg = (0, import_useTranslation.useTranslation)();
  const { Canvas } = (0, import_editor.useEditorComponents)();
  const menuCanOpen = (0, import_editor.useValue)(
    "context menu can open",
    () => !editor.getInstanceState().isCoarsePointer || editor.isIn("select"),
    [editor]
  );
  const preventEscapeFromLosingShapeFocus = (0, import_react.useCallback)(
    (e) => {
      if (e.key === "Escape") {
        e.stopPropagation();
        editor.getContainer().focus();
      }
    },
    [editor]
  );
  (0, import_react.useEffect)(() => {
    const body = editor.getContainerDocument().body;
    return () => {
      body.removeEventListener("keydown", preventEscapeFromLosingShapeFocus, {
        capture: true
      });
    };
  }, [editor, preventEscapeFromLosingShapeFocus]);
  const suppressDismissUntilRef = (0, import_react.useRef)(0);
  const cb = (0, import_react.useCallback)(
    (isOpen2) => {
      const body = editor.getContainerDocument().body;
      if (!isOpen2) {
        const onlySelectedShape = editor.getOnlySelectedShape();
        if (onlySelectedShape && editor.isShapeOrAncestorLocked(onlySelectedShape)) {
          editor.setSelectedShapes([]);
        }
        editor.timers.requestAnimationFrame(() => {
          body.removeEventListener("keydown", preventEscapeFromLosingShapeFocus, {
            capture: true
          });
        });
      } else {
        body.addEventListener("keydown", preventEscapeFromLosingShapeFocus, {
          capture: true
        });
        if (editor.getInstanceState().isCoarsePointer) {
          suppressDismissUntilRef.current = Date.now() + 500;
          const selectedShapes = editor.getSelectedShapes();
          const currentPagePoint = editor.inputs.getCurrentPagePoint();
          const shapesAtPoint = editor.getShapesAtPoint(currentPagePoint);
          if (
            // if there are no selected shapes
            !editor.getSelectedShapes().length || // OR if none of the shapes at the point include the selected shape
            !shapesAtPoint.some((s) => selectedShapes.includes(s))
          ) {
            const lockedShapes = shapesAtPoint.filter((s) => editor.isShapeOrAncestorLocked(s));
            if (lockedShapes.length) {
              editor.select(...lockedShapes.map((s) => s.id));
            }
          }
        }
      }
    },
    [editor, preventEscapeFromLosingShapeFocus]
  );
  const container = (0, import_editor.useContainer)();
  const dir = (0, import_useTranslation.useDirection)();
  const [isOpen, handleOpenChange] = (0, import_useMenuIsOpen.useMenuIsOpen)("context menu", cb);
  const content = children ?? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_DefaultContextMenuContent.DefaultContextMenuContent, {});
  return /* @__PURE__ */ (0, import_jsx_runtime.jsxs)(import_radix_ui.ContextMenu.Root, { dir, onOpenChange: handleOpenChange, modal: false, children: [
    /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_radix_ui.ContextMenu.Trigger,
      {
        onContextMenu: menuCanOpen ? void 0 : import_editor.preventDefault,
        dir: "ltr",
        disabled: disabled || !menuCanOpen,
        children: Canvas ? /* @__PURE__ */ (0, import_jsx_runtime.jsx)(Canvas, {}) : null
      }
    ),
    isOpen && /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_radix_ui.ContextMenu.Portal, { container, children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(
      import_radix_ui.ContextMenu.Content,
      {
        className: "tlui-menu tlui-scrollable",
        "data-testid": "context-menu",
        "aria-label": msg("context-menu.title"),
        alignOffset: -4,
        collisionPadding: 4,
        onContextMenu: import_editor.preventDefault,
        onPointerDownOutside: (e) => {
          if (Date.now() < suppressDismissUntilRef.current) e.preventDefault();
        },
        onInteractOutside: (e) => {
          if (Date.now() < suppressDismissUntilRef.current) e.preventDefault();
        },
        onFocusOutside: (e) => {
          if (Date.now() < suppressDismissUntilRef.current) e.preventDefault();
        },
        children: /* @__PURE__ */ (0, import_jsx_runtime.jsx)(import_TldrawUiMenuContext.TldrawUiMenuContextProvider, { type: "context-menu", sourceId: "context-menu", children: content })
      }
    ) })
  ] });
});
