// tldraw/dist-cjs/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var index_exports = {};
__export(index_exports, {
  ASPECT_RATIO_OPTIONS: () => import_crop.ASPECT_RATIO_OPTIONS,
  ASPECT_RATIO_TO_VALUE: () => import_crop.ASPECT_RATIO_TO_VALUE,
  AccessibilityMenu: () => import_AccessibilityMenu.AccessibilityMenu,
  AlignMenuItems: () => import_DefaultActionsMenuContent.AlignMenuItems,
  ArrangeMenuSubmenu: () => import_menu_items.ArrangeMenuSubmenu,
  ArrowBindingHintOverlayUtil: () => import_ArrowBindingHintOverlayUtil.ArrowBindingHintOverlayUtil,
  ArrowBindingUtil: () => import_ArrowBindingUtil.ArrowBindingUtil,
  ArrowDownToolbarItem: () => import_DefaultToolbarContent.ArrowDownToolbarItem,
  ArrowHintOverlayUtil: () => import_ArrowHintOverlayUtil.ArrowHintOverlayUtil,
  ArrowLeftToolbarItem: () => import_DefaultToolbarContent.ArrowLeftToolbarItem,
  ArrowRightToolbarItem: () => import_DefaultToolbarContent.ArrowRightToolbarItem,
  ArrowShapeTool: () => import_ArrowShapeTool.ArrowShapeTool,
  ArrowShapeUtil: () => import_ArrowShapeUtil.ArrowShapeUtil,
  ArrowToolbarItem: () => import_DefaultToolbarContent.ArrowToolbarItem,
  ArrowUpToolbarItem: () => import_DefaultToolbarContent.ArrowUpToolbarItem,
  AssetToolbarItem: () => import_DefaultToolbarContent.AssetToolbarItem,
  AssetUrlsProvider: () => import_asset_urls.AssetUrlsProvider,
  BookmarkAssetUtil: () => import_BookmarkAssetUtil.BookmarkAssetUtil,
  BookmarkShapeUtil: () => import_BookmarkShapeUtil.BookmarkShapeUtil,
  BreakPointProvider: () => import_breakpoints.BreakPointProvider,
  BrushOverlayUtil: () => import_BrushOverlayUtil.BrushOverlayUtil,
  CenteredTopPanelContainer: () => import_CenteredTopPanelContainer.CenteredTopPanelContainer,
  CheckBoxToolbarItem: () => import_DefaultToolbarContent.CheckBoxToolbarItem,
  ClipboardMenuGroup: () => import_menu_items.ClipboardMenuGroup,
  CloudToolbarItem: () => import_DefaultToolbarContent.CloudToolbarItem,
  CollaboratorBrushOverlayUtil: () => import_CollaboratorBrushOverlayUtil.CollaboratorBrushOverlayUtil,
  CollaboratorCursorOverlayUtil: () => import_CollaboratorCursorOverlayUtil.CollaboratorCursorOverlayUtil,
  CollaboratorHintOverlayUtil: () => import_CollaboratorHintOverlayUtil.CollaboratorHintOverlayUtil,
  CollaboratorScribbleOverlayUtil: () => import_CollaboratorScribbleOverlayUtil.CollaboratorScribbleOverlayUtil,
  CollaboratorShapeIndicatorOverlayUtil: () => import_CollaboratorShapeIndicatorOverlayUtil.CollaboratorShapeIndicatorOverlayUtil,
  ColorSchemeMenu: () => import_ColorSchemeMenu.ColorSchemeMenu,
  ContextMenu: () => import_DefaultContextMenu.DefaultContextMenu,
  ConversionsMenuGroup: () => import_menu_items.ConversionsMenuGroup,
  ConvertToBookmarkMenuItem: () => import_menu_items.ConvertToBookmarkMenuItem,
  ConvertToEmbedMenuItem: () => import_menu_items.ConvertToEmbedMenuItem,
  CopyAsMenuGroup: () => import_menu_items.CopyAsMenuGroup,
  CopyMenuItem: () => import_menu_items.CopyMenuItem,
  CursorChatItem: () => import_menu_items.CursorChatItem,
  CutMenuItem: () => import_menu_items.CutMenuItem,
  DEFAULT_EMBED_DEFINITIONS: () => import_defaultEmbedDefinitions.DEFAULT_EMBED_DEFINITIONS,
  DEFAULT_MAX_ASSET_SIZE: () => import_defaultExternalContentHandlers.DEFAULT_MAX_ASSET_SIZE,
  DEFAULT_MAX_IMAGE_DIMENSION: () => import_defaultExternalContentHandlers.DEFAULT_MAX_IMAGE_DIMENSION,
  DebugFlags: () => import_DefaultDebugMenuContent.DebugFlags,
  DefaultA11yAnnouncer: () => import_A11y.DefaultA11yAnnouncer,
  DefaultActionsMenu: () => import_DefaultActionsMenu.DefaultActionsMenu,
  DefaultActionsMenuContent: () => import_DefaultActionsMenuContent.DefaultActionsMenuContent,
  DefaultContextMenu: () => import_DefaultContextMenu.DefaultContextMenu,
  DefaultContextMenuContent: () => import_DefaultContextMenuContent.DefaultContextMenuContent,
  DefaultDebugMenu: () => import_DefaultDebugMenu.DefaultDebugMenu,
  DefaultDebugMenuContent: () => import_DefaultDebugMenuContent.DefaultDebugMenuContent,
  DefaultDialogs: () => import_Dialogs.DefaultDialogs,
  DefaultFollowingIndicator: () => import_DefaultFollowingIndicator.DefaultFollowingIndicator,
  DefaultFontFaces: () => import_defaultFonts.DefaultFontFaces,
  DefaultHelpMenu: () => import_DefaultHelpMenu.DefaultHelpMenu,
  DefaultHelpMenuContent: () => import_DefaultHelpMenuContent.DefaultHelpMenuContent,
  DefaultHelperButtons: () => import_DefaultHelperButtons.DefaultHelperButtons,
  DefaultHelperButtonsContent: () => import_DefaultHelperButtonsContent.DefaultHelperButtonsContent,
  DefaultImageToolbar: () => import_DefaultImageToolbar.DefaultImageToolbar,
  DefaultImageToolbarContent: () => import_DefaultImageToolbarContent.DefaultImageToolbarContent,
  DefaultKeyboardShortcutsDialog: () => import_DefaultKeyboardShortcutsDialog.DefaultKeyboardShortcutsDialog,
  DefaultKeyboardShortcutsDialogContent: () => import_DefaultKeyboardShortcutsDialogContent.DefaultKeyboardShortcutsDialogContent,
  DefaultMainMenu: () => import_DefaultMainMenu.DefaultMainMenu,
  DefaultMainMenuContent: () => import_DefaultMainMenuContent.DefaultMainMenuContent,
  DefaultMenuPanel: () => import_DefaultMenuPanel.DefaultMenuPanel,
  DefaultMinimap: () => import_DefaultMinimap.DefaultMinimap,
  DefaultNavigationPanel: () => import_DefaultNavigationPanel.DefaultNavigationPanel,
  DefaultPageMenu: () => import_DefaultPageMenu.DefaultPageMenu,
  DefaultPeopleMenu: () => import_DefaultPeopleMenu.DefaultPeopleMenu,
  DefaultPeopleMenuAvatar: () => import_DefaultPeopleMenuAvatar.DefaultPeopleMenuAvatar,
  DefaultPeopleMenuContent: () => import_DefaultPeopleMenuContent.DefaultPeopleMenuContent,
  DefaultPeopleMenuFacePile: () => import_DefaultPeopleMenuFacePile.DefaultPeopleMenuFacePile,
  DefaultPeopleMenuItem: () => import_DefaultPeopleMenuItem.DefaultPeopleMenuItem,
  DefaultQuickActions: () => import_DefaultQuickActions.DefaultQuickActions,
  DefaultQuickActionsContent: () => import_DefaultQuickActionsContent.DefaultQuickActionsContent,
  DefaultRichTextToolbar: () => import_DefaultRichTextToolbar.DefaultRichTextToolbar,
  DefaultRichTextToolbarContent: () => import_DefaultRichTextToolbarContent.DefaultRichTextToolbarContent,
  DefaultSharePanel: () => import_DefaultSharePanel.DefaultSharePanel,
  DefaultStylePanel: () => import_DefaultStylePanel.DefaultStylePanel,
  DefaultStylePanelContent: () => import_DefaultStylePanelContent.DefaultStylePanelContent,
  DefaultToasts: () => import_Toasts.DefaultToasts,
  DefaultToolbar: () => import_DefaultToolbar.DefaultToolbar,
  DefaultToolbarContent: () => import_DefaultToolbarContent.DefaultToolbarContent,
  DefaultUserPresenceEditor: () => import_DefaultUserPresenceEditor.DefaultUserPresenceEditor,
  DefaultVideoToolbar: () => import_DefaultVideoToolbar.DefaultVideoToolbar,
  DefaultVideoToolbarContent: () => import_DefaultVideoToolbarContent.DefaultVideoToolbarContent,
  DefaultZoomMenu: () => import_DefaultZoomMenu.DefaultZoomMenu,
  DefaultZoomMenuContent: () => import_DefaultZoomMenuContent.DefaultZoomMenuContent,
  DeleteMenuItem: () => import_menu_items.DeleteMenuItem,
  DiamondToolbarItem: () => import_DefaultToolbarContent.DiamondToolbarItem,
  DistributeMenuItems: () => import_DefaultActionsMenuContent.DistributeMenuItems,
  DrawShapeTool: () => import_DrawShapeTool.DrawShapeTool,
  DrawShapeUtil: () => import_DrawShapeUtil.DrawShapeUtil,
  DrawToolbarItem: () => import_DefaultToolbarContent.DrawToolbarItem,
  DuplicateMenuItem: () => import_menu_items.DuplicateMenuItem,
  EditLinkMenuItem: () => import_menu_items.EditLinkMenuItem,
  EditMenuSubmenu: () => import_menu_items.EditMenuSubmenu,
  EditSubmenu: () => import_DefaultMainMenuContent.EditSubmenu,
  EllipseToolbarItem: () => import_DefaultToolbarContent.EllipseToolbarItem,
  EmbedShapeUtil: () => import_EmbedShapeUtil.EmbedShapeUtil,
  EraserTool: () => import_EraserTool.EraserTool,
  EraserToolbarItem: () => import_DefaultToolbarContent.EraserToolbarItem,
  ExampleDialog: () => import_DefaultDebugMenuContent.ExampleDialog,
  ExportFileContentSubMenu: () => import_DefaultMainMenuContent.ExportFileContentSubMenu,
  ExtrasGroup: () => import_DefaultMainMenuContent.ExtrasGroup,
  FeatureFlags: () => import_DefaultDebugMenuContent.FeatureFlags,
  FitFrameToContentMenuItem: () => import_menu_items.FitFrameToContentMenuItem,
  FrameShapeTool: () => import_FrameShapeTool.FrameShapeTool,
  FrameShapeUtil: () => import_FrameShapeUtil.FrameShapeUtil,
  FrameToolbarItem: () => import_DefaultToolbarContent.FrameToolbarItem,
  GeoShapeTool: () => import_GeoShapeTool.GeoShapeTool,
  GeoShapeUtil: () => import_GeoShapeUtil.GeoShapeUtil,
  GroupMenuItem: () => import_menu_items.GroupMenuItem,
  GroupOrUngroupMenuItem: () => import_DefaultActionsMenuContent.GroupOrUngroupMenuItem,
  HandTool: () => import_HandTool.HandTool,
  HandToolbarItem: () => import_DefaultToolbarContent.HandToolbarItem,
  HeartToolbarItem: () => import_DefaultToolbarContent.HeartToolbarItem,
  HexagonToolbarItem: () => import_DefaultToolbarContent.HexagonToolbarItem,
  HighlightShapeTool: () => import_HighlightShapeTool.HighlightShapeTool,
  HighlightShapeUtil: () => import_HighlightShapeUtil.HighlightShapeUtil,
  HighlightToolbarItem: () => import_DefaultToolbarContent.HighlightToolbarItem,
  ImageAssetUtil: () => import_ImageAssetUtil.ImageAssetUtil,
  ImageShapeUtil: () => import_ImageShapeUtil.ImageShapeUtil,
  InputModeMenu: () => import_InputModeMenu.InputModeMenu,
  KeyboardShiftEnterTweakExtension: () => import_richText.KeyboardShiftEnterTweakExtension,
  KeyboardShortcutsMenuItem: () => import_DefaultHelpMenuContent.KeyboardShortcutsMenuItem,
  LanguageMenu: () => import_LanguageMenu.LanguageMenu,
  LaserTool: () => import_LaserTool.LaserTool,
  LaserToolbarItem: () => import_DefaultToolbarContent.LaserToolbarItem,
  LineShapeTool: () => import_LineShapeTool.LineShapeTool,
  LineShapeUtil: () => import_LineShapeUtil.LineShapeUtil,
  LineToolbarItem: () => import_DefaultToolbarContent.LineToolbarItem,
  LockGroup: () => import_DefaultMainMenuContent.LockGroup,
  MiscMenuGroup: () => import_DefaultMainMenuContent.MiscMenuGroup,
  MobileStylePanel: () => import_MobileStylePanel.MobileStylePanel,
  MoveToPageMenu: () => import_menu_items.MoveToPageMenu,
  NoteShapeTool: () => import_NoteShapeTool.NoteShapeTool,
  NoteShapeUtil: () => import_NoteShapeUtil.NoteShapeUtil,
  NoteToolbarItem: () => import_DefaultToolbarContent.NoteToolbarItem,
  OfflineIndicator: () => import_OfflineIndicator.OfflineIndicator,
  OvalToolbarItem: () => import_DefaultToolbarContent.OvalToolbarItem,
  OverflowingToolbar: () => import_OverflowingToolbar.OverflowingToolbar,
  PORTRAIT_BREAKPOINT: () => import_constants.PORTRAIT_BREAKPOINT,
  PageItemInput: () => import_PageItemInput.PageItemInput,
  PageItemSubmenu: () => import_PageItemSubmenu.PageItemSubmenu,
  PasteMenuItem: () => import_menu_items.PasteMenuItem,
  PathBuilder: () => import_PathBuilder.PathBuilder,
  PathBuilderGeometry2d: () => import_PathBuilder.PathBuilderGeometry2d,
  PeopleMenu: () => import_DefaultPeopleMenu.DefaultPeopleMenu,
  PlainTextArea: () => import_PlainTextArea.PlainTextArea,
  PlainTextLabel: () => import_PlainTextLabel.PlainTextLabel,
  PreferencesGroup: () => import_DefaultMainMenuContent.PreferencesGroup,
  PrintItem: () => import_menu_items.PrintItem,
  RTL_LANGUAGES: () => import_translations.RTL_LANGUAGES,
  RectangleToolbarItem: () => import_DefaultToolbarContent.RectangleToolbarItem,
  RemoveFrameMenuItem: () => import_menu_items.RemoveFrameMenuItem,
  ReorderMenuItems: () => import_DefaultActionsMenuContent.ReorderMenuItems,
  ReorderMenuSubmenu: () => import_menu_items.ReorderMenuSubmenu,
  RhombusToolbarItem: () => import_DefaultToolbarContent.RhombusToolbarItem,
  RichTextArea: () => import_RichTextArea.RichTextArea,
  RichTextLabel: () => import_RichTextLabel.RichTextLabel,
  RichTextSVG: () => import_RichTextLabel.RichTextSVG,
  RotateCWMenuItem: () => import_DefaultActionsMenuContent.RotateCWMenuItem,
  ScribbleOverlayUtil: () => import_ScribbleOverlayUtil.ScribbleOverlayUtil,
  SelectAllMenuItem: () => import_menu_items.SelectAllMenuItem,
  SelectTool: () => import_SelectTool.SelectTool,
  SelectToolbarItem: () => import_DefaultToolbarContent.SelectToolbarItem,
  SelectionForegroundOverlayUtil: () => import_SelectionForegroundOverlayUtil.SelectionForegroundOverlayUtil,
  ShapeHandleOverlayUtil: () => import_ShapeHandleOverlayUtil.ShapeHandleOverlayUtil,
  ShapeIndicatorOverlayUtil: () => import_ShapeIndicatorOverlayUtil.ShapeIndicatorOverlayUtil,
  SnapIndicatorOverlayUtil: () => import_SnapIndicatorOverlayUtil.SnapIndicatorOverlayUtil,
  Spinner: () => import_Spinner.Spinner,
  StackMenuItems: () => import_DefaultActionsMenuContent.StackMenuItems,
  StarToolbarItem: () => import_DefaultToolbarContent.StarToolbarItem,
  StylePanelArrowKindPicker: () => import_DefaultStylePanelContent.StylePanelArrowKindPicker,
  StylePanelArrowheadPicker: () => import_DefaultStylePanelContent.StylePanelArrowheadPicker,
  StylePanelButtonPicker: () => import_StylePanelButtonPicker.StylePanelButtonPicker,
  StylePanelButtonPickerInline: () => import_StylePanelButtonPicker.StylePanelButtonPickerInline,
  StylePanelColorPicker: () => import_DefaultStylePanelContent.StylePanelColorPicker,
  StylePanelContextProvider: () => import_StylePanelContext.StylePanelContextProvider,
  StylePanelDashPicker: () => import_DefaultStylePanelContent.StylePanelDashPicker,
  StylePanelDoubleDropdownPicker: () => import_StylePanelDoubleDropdownPicker.StylePanelDoubleDropdownPicker,
  StylePanelDoubleDropdownPickerInline: () => import_StylePanelDoubleDropdownPicker.StylePanelDoubleDropdownPickerInline,
  StylePanelDropdownPicker: () => import_StylePanelDropdownPicker.StylePanelDropdownPicker,
  StylePanelDropdownPickerInline: () => import_StylePanelDropdownPicker.StylePanelDropdownPickerInline,
  StylePanelFillPicker: () => import_DefaultStylePanelContent.StylePanelFillPicker,
  StylePanelFontPicker: () => import_DefaultStylePanelContent.StylePanelFontPicker,
  StylePanelGeoShapePicker: () => import_DefaultStylePanelContent.StylePanelGeoShapePicker,
  StylePanelLabelAlignPicker: () => import_DefaultStylePanelContent.StylePanelLabelAlignPicker,
  StylePanelOpacityPicker: () => import_DefaultStylePanelContent.StylePanelOpacityPicker,
  StylePanelSection: () => import_DefaultStylePanelContent.StylePanelSection,
  StylePanelSizePicker: () => import_DefaultStylePanelContent.StylePanelSizePicker,
  StylePanelSplinePicker: () => import_DefaultStylePanelContent.StylePanelSplinePicker,
  StylePanelSubheading: () => import_StylePanelSubheading.StylePanelSubheading,
  StylePanelTextAlignPicker: () => import_DefaultStylePanelContent.StylePanelTextAlignPicker,
  TLDRAW_FILE_EXTENSION: () => import_file.TLDRAW_FILE_EXTENSION,
  TLV1AlignStyle: () => import_buildFromV1Document.TLV1AlignStyle,
  TLV1AssetType: () => import_buildFromV1Document.TLV1AssetType,
  TLV1ColorStyle: () => import_buildFromV1Document.TLV1ColorStyle,
  TLV1DashStyle: () => import_buildFromV1Document.TLV1DashStyle,
  TLV1Decoration: () => import_buildFromV1Document.TLV1Decoration,
  TLV1FontStyle: () => import_buildFromV1Document.TLV1FontStyle,
  TLV1ShapeType: () => import_buildFromV1Document.TLV1ShapeType,
  TLV1SizeStyle: () => import_buildFromV1Document.TLV1SizeStyle,
  TextShapeTool: () => import_TextShapeTool.TextShapeTool,
  TextShapeUtil: () => import_TextShapeUtil.TextShapeUtil,
  TextToolbarItem: () => import_DefaultToolbarContent.TextToolbarItem,
  Tldraw: () => import_Tldraw.Tldraw,
  TldrawImage: () => import_TldrawImage.TldrawImage,
  TldrawUi: () => import_TldrawUi.TldrawUi,
  TldrawUiA11yProvider: () => import_a11y.TldrawUiA11yProvider,
  TldrawUiButton: () => import_TldrawUiButton.TldrawUiButton,
  TldrawUiButtonCheck: () => import_TldrawUiButtonCheck.TldrawUiButtonCheck,
  TldrawUiButtonIcon: () => import_TldrawUiButtonIcon.TldrawUiButtonIcon,
  TldrawUiButtonLabel: () => import_TldrawUiButtonLabel.TldrawUiButtonLabel,
  TldrawUiColumn: () => import_layout.TldrawUiColumn,
  TldrawUiComponentsProvider: () => import_components.TldrawUiComponentsProvider,
  TldrawUiContextProvider: () => import_TldrawUiContextProvider.TldrawUiContextProvider,
  TldrawUiContextualToolbar: () => import_TldrawUiContextualToolbar.TldrawUiContextualToolbar,
  TldrawUiDialogBody: () => import_TldrawUiDialog.TldrawUiDialogBody,
  TldrawUiDialogCloseButton: () => import_TldrawUiDialog.TldrawUiDialogCloseButton,
  TldrawUiDialogFooter: () => import_TldrawUiDialog.TldrawUiDialogFooter,
  TldrawUiDialogHeader: () => import_TldrawUiDialog.TldrawUiDialogHeader,
  TldrawUiDialogTitle: () => import_TldrawUiDialog.TldrawUiDialogTitle,
  TldrawUiDialogsProvider: () => import_dialogs.TldrawUiDialogsProvider,
  TldrawUiDropdownMenuCheckboxItem: () => import_TldrawUiDropdownMenu.TldrawUiDropdownMenuCheckboxItem,
  TldrawUiDropdownMenuContent: () => import_TldrawUiDropdownMenu.TldrawUiDropdownMenuContent,
  TldrawUiDropdownMenuGroup: () => import_TldrawUiDropdownMenu.TldrawUiDropdownMenuGroup,
  TldrawUiDropdownMenuIndicator: () => import_TldrawUiDropdownMenu.TldrawUiDropdownMenuIndicator,
  TldrawUiDropdownMenuItem: () => import_TldrawUiDropdownMenu.TldrawUiDropdownMenuItem,
  TldrawUiDropdownMenuRoot: () => import_TldrawUiDropdownMenu.TldrawUiDropdownMenuRoot,
  TldrawUiDropdownMenuSub: () => import_TldrawUiDropdownMenu.TldrawUiDropdownMenuSub,
  TldrawUiDropdownMenuSubContent: () => import_TldrawUiDropdownMenu.TldrawUiDropdownMenuSubContent,
  TldrawUiDropdownMenuSubTrigger: () => import_TldrawUiDropdownMenu.TldrawUiDropdownMenuSubTrigger,
  TldrawUiDropdownMenuTrigger: () => import_TldrawUiDropdownMenu.TldrawUiDropdownMenuTrigger,
  TldrawUiEventsProvider: () => import_events.TldrawUiEventsProvider,
  TldrawUiGrid: () => import_layout.TldrawUiGrid,
  TldrawUiIcon: () => import_TldrawUiIcon.TldrawUiIcon,
  TldrawUiInFrontOfTheCanvas: () => import_TldrawUi.TldrawUiInFrontOfTheCanvas,
  TldrawUiInput: () => import_TldrawUiInput.TldrawUiInput,
  TldrawUiKbd: () => import_TldrawUiKbd.TldrawUiKbd,
  TldrawUiMenuActionCheckboxItem: () => import_TldrawUiMenuActionCheckboxItem.TldrawUiMenuActionCheckboxItem,
  TldrawUiMenuActionItem: () => import_TldrawUiMenuActionItem.TldrawUiMenuActionItem,
  TldrawUiMenuCheckboxItem: () => import_TldrawUiMenuCheckboxItem.TldrawUiMenuCheckboxItem,
  TldrawUiMenuContextProvider: () => import_TldrawUiMenuContext.TldrawUiMenuContextProvider,
  TldrawUiMenuGroup: () => import_TldrawUiMenuGroup.TldrawUiMenuGroup,
  TldrawUiMenuItem: () => import_TldrawUiMenuItem.TldrawUiMenuItem,
  TldrawUiMenuSubmenu: () => import_TldrawUiMenuSubmenu.TldrawUiMenuSubmenu,
  TldrawUiMenuToolItem: () => import_TldrawUiMenuToolItem.TldrawUiMenuToolItem,
  TldrawUiOrientationProvider: () => import_layout.TldrawUiOrientationProvider,
  TldrawUiPopover: () => import_TldrawUiPopover.TldrawUiPopover,
  TldrawUiPopoverContent: () => import_TldrawUiPopover.TldrawUiPopoverContent,
  TldrawUiPopoverTrigger: () => import_TldrawUiPopover.TldrawUiPopoverTrigger,
  TldrawUiRow: () => import_layout.TldrawUiRow,
  TldrawUiSelect: () => import_TldrawUiSelect.TldrawUiSelect,
  TldrawUiSelectContent: () => import_TldrawUiSelect.TldrawUiSelectContent,
  TldrawUiSelectItem: () => import_TldrawUiSelect.TldrawUiSelectItem,
  TldrawUiSelectTrigger: () => import_TldrawUiSelect.TldrawUiSelectTrigger,
  TldrawUiSelectValue: () => import_TldrawUiSelect.TldrawUiSelectValue,
  TldrawUiSlider: () => import_TldrawUiSlider.TldrawUiSlider,
  TldrawUiToastsProvider: () => import_toasts.TldrawUiToastsProvider,
  TldrawUiToolbar: () => import_TldrawUiToolbar.TldrawUiToolbar,
  TldrawUiToolbarButton: () => import_TldrawUiToolbar.TldrawUiToolbarButton,
  TldrawUiToolbarToggleGroup: () => import_TldrawUiToolbar.TldrawUiToolbarToggleGroup,
  TldrawUiToolbarToggleItem: () => import_TldrawUiToolbar.TldrawUiToolbarToggleItem,
  TldrawUiTooltip: () => import_TldrawUiTooltip.TldrawUiTooltip,
  TldrawUiTooltipProvider: () => import_TldrawUiTooltip.TldrawUiTooltipProvider,
  TldrawUiTranslationProvider: () => import_useTranslation.TldrawUiTranslationProvider,
  ToggleAutoSizeMenuItem: () => import_menu_items.ToggleAutoSizeMenuItem,
  ToggleDebugModeItem: () => import_menu_items.ToggleDebugModeItem,
  ToggleDynamicSizeModeItem: () => import_menu_items.ToggleDynamicSizeModeItem,
  ToggleEdgeScrollingItem: () => import_menu_items.ToggleEdgeScrollingItem,
  ToggleEnhancedA11yModeItem: () => import_menu_items.ToggleEnhancedA11yModeItem,
  ToggleFocusModeItem: () => import_menu_items.ToggleFocusModeItem,
  ToggleGridItem: () => import_menu_items.ToggleGridItem,
  ToggleInvertZoomItem: () => import_menu_items.ToggleInvertZoomItem,
  ToggleKeyboardShortcutsItem: () => import_menu_items.ToggleKeyboardShortcutsItem,
  ToggleLockMenuItem: () => import_menu_items.ToggleLockMenuItem,
  TogglePasteAtCursorItem: () => import_menu_items.TogglePasteAtCursorItem,
  ToggleReduceMotionItem: () => import_menu_items.ToggleReduceMotionItem,
  ToggleSnapModeItem: () => import_menu_items.ToggleSnapModeItem,
  ToggleToolLockItem: () => import_menu_items.ToggleToolLockItem,
  ToggleToolLockedButton: () => import_ToggleToolLockedButton.ToggleToolLockedButton,
  ToggleTransparentBgMenuItem: () => import_menu_items.ToggleTransparentBgMenuItem,
  ToggleWrapModeItem: () => import_menu_items.ToggleWrapModeItem,
  ToolbarItem: () => import_DefaultToolbarContent.ToolbarItem,
  TrapezoidToolbarItem: () => import_DefaultToolbarContent.TrapezoidToolbarItem,
  TriangleToolbarItem: () => import_DefaultToolbarContent.TriangleToolbarItem,
  UndoRedoGroup: () => import_DefaultMainMenuContent.UndoRedoGroup,
  UngroupMenuItem: () => import_menu_items.UngroupMenuItem,
  UnlockAllMenuItem: () => import_menu_items.UnlockAllMenuItem,
  VideoAssetUtil: () => import_VideoAssetUtil.VideoAssetUtil,
  VideoShapeUtil: () => import_VideoShapeUtil.VideoShapeUtil,
  ViewSubmenu: () => import_DefaultMainMenuContent.ViewSubmenu,
  XBoxToolbarItem: () => import_DefaultToolbarContent.XBoxToolbarItem,
  ZoomBrushOverlayUtil: () => import_ZoomBrushOverlayUtil.ZoomBrushOverlayUtil,
  ZoomOrRotateMenuItem: () => import_DefaultActionsMenuContent.ZoomOrRotateMenuItem,
  ZoomTo100MenuItem: () => import_menu_items.ZoomTo100MenuItem,
  ZoomToFitMenuItem: () => import_menu_items.ZoomToFitMenuItem,
  ZoomToSelectionMenuItem: () => import_menu_items.ZoomToSelectionMenuItem,
  ZoomTool: () => import_ZoomTool.ZoomTool,
  allDefaultFontFaces: () => import_defaultFonts.allDefaultFontFaces,
  buildFromV1Document: () => import_buildFromV1Document.buildFromV1Document,
  centerSelectionAroundPoint: () => import_defaultExternalContentHandlers.centerSelectionAroundPoint,
  clearArrowTargetState: () => import_arrowTargetState.clearArrowTargetState,
  containBoxSize: () => import_assets.containBoxSize,
  copyAs: () => import_copyAs.copyAs,
  createBookmarkFromUrl: () => import_bookmarks.createBookmarkFromUrl,
  createEmptyBookmarkShape: () => import_defaultExternalContentHandlers.createEmptyBookmarkShape,
  createShapesForAssets: () => import_defaultExternalContentHandlers.createShapesForAssets,
  defaultAddFontsFromNode: () => import_richText.defaultAddFontsFromNode,
  defaultAssetUtils: () => import_defaultAssetUtils.defaultAssetUtils,
  defaultBindingUtils: () => import_defaultBindingUtils.defaultBindingUtils,
  defaultEditorAssetUrls: () => import_assetUrls2.defaultEditorAssetUrls,
  defaultGeoTypeDefinitions: () => import_getGeoShapePath.defaultGeoTypeDefinitions,
  defaultHandleExternalEmbedContent: () => import_defaultExternalContentHandlers.defaultHandleExternalEmbedContent,
  defaultHandleExternalExcalidrawContent: () => import_defaultExternalContentHandlers.defaultHandleExternalExcalidrawContent,
  defaultHandleExternalFileAsset: () => import_defaultExternalContentHandlers.defaultHandleExternalFileAsset,
  defaultHandleExternalFileContent: () => import_defaultExternalContentHandlers.defaultHandleExternalFileContent,
  defaultHandleExternalFileReplaceContent: () => import_defaultExternalContentHandlers.defaultHandleExternalFileReplaceContent,
  defaultHandleExternalSvgTextContent: () => import_defaultExternalContentHandlers.defaultHandleExternalSvgTextContent,
  defaultHandleExternalTextContent: () => import_defaultExternalContentHandlers.defaultHandleExternalTextContent,
  defaultHandleExternalTldrawContent: () => import_defaultExternalContentHandlers.defaultHandleExternalTldrawContent,
  defaultHandleExternalUrlAsset: () => import_defaultExternalContentHandlers.defaultHandleExternalUrlAsset,
  defaultHandleExternalUrlContent: () => import_defaultExternalContentHandlers.defaultHandleExternalUrlContent,
  defaultOverlayUtils: () => import_defaultOverlayUtils.defaultOverlayUtils,
  defaultShapeTools: () => import_defaultShapeTools.defaultShapeTools,
  defaultShapeUtils: () => import_defaultShapeUtils.defaultShapeUtils,
  defaultTools: () => import_defaultTools.defaultTools,
  downloadFile: () => import_exportAs.downloadFile,
  downsizeImage: () => import_assets.downsizeImage,
  embedShapePermissionDefaults: () => import_defaultEmbedDefinitions.embedShapePermissionDefaults,
  exportAs: () => import_exportAs.exportAs,
  fitFrameToContent: () => import_frames.fitFrameToContent,
  getArrowBindings: () => import_shared.getArrowBindings,
  getArrowInfo: () => import_getArrowInfo.getArrowInfo,
  getArrowTargetState: () => import_arrowTargetState.getArrowTargetState,
  getArrowTerminalsInArrowSpace: () => import_shared.getArrowTerminalsInArrowSpace,
  getAssetInfo: () => import_defaultExternalContentHandlers.getAssetInfo,
  getColorStyleItems: () => import_styles.getColorStyleItems,
  getCropBox: () => import_crop.getCropBox,
  getDefaultCrop: () => import_crop.getDefaultCrop,
  getDisplayValues: () => import_getDisplayValues.getDisplayValues,
  getEmbedInfo: () => import_embeds.getEmbedInfo,
  getFontFamily: () => import_default_shape_constants.getFontFamily,
  getFontStyleItems: () => import_styles.getFontStyleItems,
  getGeoTypeDefinition: () => import_getGeoShapePath.getGeoTypeDefinition,
  getHitShapeOnCanvasPointerDown: () => import_getHitShapeOnCanvasPointerDown.getHitShapeOnCanvasPointerDown,
  getPointsFromDrawSegment: () => import_getPath.getPointsFromDrawSegment,
  getPointsFromDrawSegments: () => import_getPath.getPointsFromDrawSegments,
  getStroke: () => import_getStroke.getStroke,
  getStrokeOutlinePoints: () => import_getStrokeOutlinePoints.getStrokeOutlinePoints,
  getStrokePoints: () => import_getStrokePoints.getStrokePoints,
  getSvgPathFromStrokePoints: () => import_svg.getSvgPathFromStrokePoints,
  getUncroppedSize: () => import_crop.getUncroppedSize,
  handleNativeOrMenuCopy: () => import_useClipboardEvents.handleNativeOrMenuCopy,
  hideAllTooltips: () => import_TldrawUiTooltip.hideAllTooltips,
  iconTypes: () => import_icon_types.iconTypes,
  notifyIfFileNotAllowed: () => import_defaultExternalContentHandlers.notifyIfFileNotAllowed,
  onDragFromToolbarToCreateShape: () => import_useTools.onDragFromToolbarToCreateShape,
  parseAndLoadDocument: () => import_file.parseAndLoadDocument,
  parseTldrawJsonFile: () => import_file.parseTldrawJsonFile,
  preloadFont: () => import_preload_font.preloadFont,
  putExcalidrawContent: () => import_putExcalidrawContent.putExcalidrawContent,
  registerDefaultExternalContentHandlers: () => import_defaultExternalContentHandlers.registerDefaultExternalContentHandlers,
  registerDefaultSideEffects: () => import_defaultSideEffects.registerDefaultSideEffects,
  removeFrame: () => import_frames.removeFrame,
  renderHtmlFromRichText: () => import_richText.renderHtmlFromRichText,
  renderHtmlFromRichTextForMeasurement: () => import_richText.renderHtmlFromRichTextForMeasurement,
  renderPlaintextFromRichText: () => import_richText.renderPlaintextFromRichText,
  renderRichTextFromHTML: () => import_richText.renderRichTextFromHTML,
  sanitizeSvg: () => import_sanitizeSvg.sanitizeSvg,
  serializeTldrawJson: () => import_file.serializeTldrawJson,
  serializeTldrawJsonBlob: () => import_file.serializeTldrawJsonBlob,
  setDefaultEditorAssetUrls: () => import_assetUrls2.setDefaultEditorAssetUrls,
  setDefaultUiAssetUrls: () => import_assetUrls.setDefaultUiAssetUrls,
  startEditingShapeWithRichText: () => import_selectHelpers.startEditingShapeWithRichText,
  tipTapDefaultExtensions: () => import_richText.tipTapDefaultExtensions,
  truncateStringWithEllipsis: () => import_text.truncateStringWithEllipsis,
  unknownEmbedShapePermissionOverrides: () => import_defaultEmbedDefinitions.unknownEmbedShapePermissionOverrides,
  unwrapLabel: () => import_actions.unwrapLabel,
  updateArrowTargetState: () => import_arrowTargetState.updateArrowTargetState,
  useA11y: () => import_a11y.useA11y,
  useActions: () => import_actions.useActions,
  useAssetUrls: () => import_asset_urls.useAssetUrls,
  useBreakpoint: () => import_breakpoints.useBreakpoint,
  useCanApplySelectionAction: () => import_menu_hooks.useCanApplySelectionAction,
  useCanRedo: () => import_menu_hooks.useCanRedo,
  useCanUndo: () => import_menu_hooks.useCanUndo,
  useCollaborationStatus: () => import_useCollaborationStatus.useCollaborationStatus,
  useCopyAs: () => import_useCopyAs.useCopyAs,
  useCurrentTranslation: () => import_useTranslation2.useCurrentTranslation,
  useDefaultHelpers: () => import_overrides.useDefaultHelpers,
  useDialogs: () => import_dialogs.useDialogs,
  useDirection: () => import_useTranslation2.useDirection,
  useEditablePlainText: () => import_useEditablePlainText.useEditablePlainText,
  useEditableRichText: () => import_useEditableRichText.useEditableRichText,
  useExportAs: () => import_useExportAs.useExportAs,
  useImageOrVideoAsset: () => import_useImageOrVideoAsset.useImageOrVideoAsset,
  useIsToolSelected: () => import_DefaultToolbarContent.useIsToolSelected,
  useKeyboardShortcuts: () => import_useKeyboardShortcuts.useKeyboardShortcuts,
  useLocalStorageState: () => import_useLocalStorageState.useLocalStorageState,
  useMenuClipboardEvents: () => import_useClipboardEvents.useMenuClipboardEvents,
  useMenuIsOpen: () => import_useMenuIsOpen.useMenuIsOpen,
  useNativeClipboardEvents: () => import_useClipboardEvents.useNativeClipboardEvents,
  usePrefersReducedMotion: () => import_usePrefersReducedMotion.usePrefersReducedMotion,
  useReadonly: () => import_useReadonly.useReadonly,
  useRelevantStyles: () => import_useRelevantStyles.useRelevantStyles,
  useSelectedShapesAnnouncer: () => import_A11y.useSelectedShapesAnnouncer,
  useShowCollaborationUi: () => import_useCollaborationStatus.useShowCollaborationUi,
  useStylePanelContext: () => import_StylePanelContext.useStylePanelContext,
  useTldrawUiComponents: () => import_components.useTldrawUiComponents,
  useTldrawUiOrientation: () => import_layout.useTldrawUiOrientation,
  useToasts: () => import_toasts.useToasts,
  useTools: () => import_useTools.useTools,
  useTranslation: () => import_useTranslation2.useTranslation,
  useUiEvents: () => import_events.useUiEvents,
  useUnlockedSelectedShapesCount: () => import_menu_hooks.useUnlockedSelectedShapesCount
});
module.exports = __toCommonJS(index_exports);
var import_editor = require("@tldraw/editor");
var import_getPath = require("./lib/shapes/draw/getPath");
var import_PathBuilder = require("./lib/shapes/shared/PathBuilder");
var import_usePrefersReducedMotion = require("./lib/shapes/shared/usePrefersReducedMotion");
var import_selectHelpers = require("./lib/tools/SelectTool/selectHelpers");
var import_A11y = require("./lib/ui/components/A11y");
var import_AccessibilityMenu = require("./lib/ui/components/AccessibilityMenu");
var import_ColorSchemeMenu = require("./lib/ui/components/ColorSchemeMenu");
var import_DefaultFollowingIndicator = require("./lib/ui/components/DefaultFollowingIndicator");
var import_Dialogs = require("./lib/ui/components/Dialogs");
var import_layout = require("./lib/ui/components/primitives/layout");
var import_TldrawUiMenuActionCheckboxItem = require("./lib/ui/components/primitives/menus/TldrawUiMenuActionCheckboxItem");
var import_TldrawUiMenuActionItem = require("./lib/ui/components/primitives/menus/TldrawUiMenuActionItem");
var import_TldrawUiMenuToolItem = require("./lib/ui/components/primitives/menus/TldrawUiMenuToolItem");
var import_Toasts = require("./lib/ui/components/Toasts");
var import_useTranslation = require("./lib/ui/hooks/useTranslation/useTranslation");
__reExport(index_exports, require("@tldraw/editor"), module.exports);
var import_BookmarkAssetUtil = require("./lib/assets/BookmarkAssetUtil");
var import_ImageAssetUtil = require("./lib/assets/ImageAssetUtil");
var import_VideoAssetUtil = require("./lib/assets/VideoAssetUtil");
var import_ArrowBindingHintOverlayUtil = require("./lib/overlays/ArrowBindingHintOverlayUtil");
var import_ArrowBindingUtil = require("./lib/bindings/arrow/ArrowBindingUtil");
var import_ArrowHintOverlayUtil = require("./lib/overlays/ArrowHintOverlayUtil");
var import_BrushOverlayUtil = require("./lib/overlays/BrushOverlayUtil");
var import_ZoomBrushOverlayUtil = require("./lib/overlays/ZoomBrushOverlayUtil");
var import_ScribbleOverlayUtil = require("./lib/overlays/ScribbleOverlayUtil");
var import_CollaboratorBrushOverlayUtil = require("./lib/overlays/CollaboratorBrushOverlayUtil");
var import_CollaboratorScribbleOverlayUtil = require("./lib/overlays/CollaboratorScribbleOverlayUtil");
var import_CollaboratorHintOverlayUtil = require("./lib/overlays/CollaboratorHintOverlayUtil");
var import_CollaboratorCursorOverlayUtil = require("./lib/overlays/CollaboratorCursorOverlayUtil");
var import_CollaboratorShapeIndicatorOverlayUtil = require("./lib/overlays/CollaboratorShapeIndicatorOverlayUtil");
var import_ShapeHandleOverlayUtil = require("./lib/overlays/ShapeHandleOverlayUtil");
var import_ShapeIndicatorOverlayUtil = require("./lib/overlays/ShapeIndicatorOverlayUtil");
var import_SelectionForegroundOverlayUtil = require("./lib/overlays/SelectionForegroundOverlayUtil");
var import_SnapIndicatorOverlayUtil = require("./lib/overlays/SnapIndicatorOverlayUtil");
var import_defaultAssetUtils = require("./lib/defaultAssetUtils");
var import_defaultBindingUtils = require("./lib/defaultBindingUtils");
var import_defaultOverlayUtils = require("./lib/defaultOverlayUtils");
var import_defaultEmbedDefinitions = require("./lib/defaultEmbedDefinitions");
var import_defaultExternalContentHandlers = require("./lib/defaultExternalContentHandlers");
var import_defaultShapeTools = require("./lib/defaultShapeTools");
var import_defaultShapeUtils = require("./lib/defaultShapeUtils");
var import_defaultSideEffects = require("./lib/defaultSideEffects");
var import_defaultTools = require("./lib/defaultTools");
var import_ArrowShapeTool = require("./lib/shapes/arrow/ArrowShapeTool");
var import_ArrowShapeUtil = require("./lib/shapes/arrow/ArrowShapeUtil");
var import_arrowTargetState = require("./lib/shapes/arrow/arrowTargetState");
var import_getArrowInfo = require("./lib/shapes/arrow/getArrowInfo");
var import_shared = require("./lib/shapes/arrow/shared");
var import_bookmarks = require("./lib/shapes/bookmark/bookmarks");
var import_BookmarkShapeUtil = require("./lib/shapes/bookmark/BookmarkShapeUtil");
var import_DrawShapeTool = require("./lib/shapes/draw/DrawShapeTool");
var import_DrawShapeUtil = require("./lib/shapes/draw/DrawShapeUtil");
var import_EmbedShapeUtil = require("./lib/shapes/embed/EmbedShapeUtil");
var import_FrameShapeTool = require("./lib/shapes/frame/FrameShapeTool");
var import_FrameShapeUtil = require("./lib/shapes/frame/FrameShapeUtil");
var import_GeoShapeTool = require("./lib/shapes/geo/GeoShapeTool");
var import_GeoShapeUtil = require("./lib/shapes/geo/GeoShapeUtil");
var import_getGeoShapePath = require("./lib/shapes/geo/getGeoShapePath");
var import_HighlightShapeTool = require("./lib/shapes/highlight/HighlightShapeTool");
var import_HighlightShapeUtil = require("./lib/shapes/highlight/HighlightShapeUtil");
var import_ImageShapeUtil = require("./lib/shapes/image/ImageShapeUtil");
var import_LineShapeTool = require("./lib/shapes/line/LineShapeTool");
var import_LineShapeUtil = require("./lib/shapes/line/LineShapeUtil");
var import_NoteShapeTool = require("./lib/shapes/note/NoteShapeTool");
var import_NoteShapeUtil = require("./lib/shapes/note/NoteShapeUtil");
var import_crop = require("./lib/shapes/shared/crop");
var import_default_shape_constants = require("./lib/shapes/shared/default-shape-constants");
var import_defaultFonts = require("./lib/shapes/shared/defaultFonts");
var import_getStroke = require("./lib/shapes/shared/freehand/getStroke");
var import_getStrokeOutlinePoints = require("./lib/shapes/shared/freehand/getStrokeOutlinePoints");
var import_getStrokePoints = require("./lib/shapes/shared/freehand/getStrokePoints");
var import_svg = require("./lib/shapes/shared/freehand/svg");
var import_getDisplayValues = require("./lib/shapes/shared/getDisplayValues");
var import_PlainTextLabel = require("./lib/shapes/shared/PlainTextLabel");
var import_RichTextLabel = require("./lib/shapes/shared/RichTextLabel");
var import_useEditablePlainText = require("./lib/shapes/shared/useEditablePlainText");
var import_useEditableRichText = require("./lib/shapes/shared/useEditableRichText");
var import_useImageOrVideoAsset = require("./lib/shapes/shared/useImageOrVideoAsset");
var import_PlainTextArea = require("./lib/shapes/text/PlainTextArea");
var import_RichTextArea = require("./lib/shapes/text/RichTextArea");
var import_TextShapeTool = require("./lib/shapes/text/TextShapeTool");
var import_TextShapeUtil = require("./lib/shapes/text/TextShapeUtil");
var import_VideoShapeUtil = require("./lib/shapes/video/VideoShapeUtil");
var import_styles = require("./lib/styles");
var import_Tldraw = require("./lib/Tldraw");
var import_TldrawImage = require("./lib/TldrawImage");
var import_EraserTool = require("./lib/tools/EraserTool/EraserTool");
var import_HandTool = require("./lib/tools/HandTool/HandTool");
var import_LaserTool = require("./lib/tools/LaserTool/LaserTool");
var import_getHitShapeOnCanvasPointerDown = require("./lib/tools/selection-logic/getHitShapeOnCanvasPointerDown");
var import_SelectTool = require("./lib/tools/SelectTool/SelectTool");
var import_ZoomTool = require("./lib/tools/ZoomTool/ZoomTool");
var import_assetUrls = require("./lib/ui/assetUrls");
var import_DefaultActionsMenu = require("./lib/ui/components/ActionsMenu/DefaultActionsMenu");
var import_DefaultActionsMenuContent = require("./lib/ui/components/ActionsMenu/DefaultActionsMenuContent");
var import_DefaultContextMenu = require("./lib/ui/components/ContextMenu/DefaultContextMenu");
var import_DefaultContextMenuContent = require("./lib/ui/components/ContextMenu/DefaultContextMenuContent");
var import_DefaultDebugMenu = require("./lib/ui/components/DebugMenu/DefaultDebugMenu");
var import_DefaultDebugMenuContent = require("./lib/ui/components/DebugMenu/DefaultDebugMenuContent");
var import_DefaultMenuPanel = require("./lib/ui/components/DefaultMenuPanel");
var import_DefaultHelperButtons = require("./lib/ui/components/HelperButtons/DefaultHelperButtons");
var import_DefaultHelperButtonsContent = require("./lib/ui/components/HelperButtons/DefaultHelperButtonsContent");
var import_DefaultHelpMenu = require("./lib/ui/components/HelpMenu/DefaultHelpMenu");
var import_DefaultHelpMenuContent = require("./lib/ui/components/HelpMenu/DefaultHelpMenuContent");
var import_DefaultKeyboardShortcutsDialog = require("./lib/ui/components/KeyboardShortcutsDialog/DefaultKeyboardShortcutsDialog");
var import_DefaultKeyboardShortcutsDialogContent = require("./lib/ui/components/KeyboardShortcutsDialog/DefaultKeyboardShortcutsDialogContent");
var import_LanguageMenu = require("./lib/ui/components/LanguageMenu");
var import_InputModeMenu = require("./lib/ui/components/InputModeMenu");
var import_DefaultMainMenu = require("./lib/ui/components/MainMenu/DefaultMainMenu");
var import_DefaultMainMenuContent = require("./lib/ui/components/MainMenu/DefaultMainMenuContent");
var import_menu_items = require("./lib/ui/components/menu-items");
var import_DefaultMinimap = require("./lib/ui/components/Minimap/DefaultMinimap");
var import_MobileStylePanel = require("./lib/ui/components/MobileStylePanel");
var import_DefaultNavigationPanel = require("./lib/ui/components/NavigationPanel/DefaultNavigationPanel");
var import_OfflineIndicator = require("./lib/ui/components/OfflineIndicator/OfflineIndicator");
var import_DefaultPageMenu = require("./lib/ui/components/PageMenu/DefaultPageMenu");
var import_PageItemInput = require("./lib/ui/components/PageMenu/PageItemInput");
var import_PageItemSubmenu = require("./lib/ui/components/PageMenu/PageItemSubmenu");
var import_TldrawUiButton = require("./lib/ui/components/primitives/Button/TldrawUiButton");
var import_TldrawUiButtonCheck = require("./lib/ui/components/primitives/Button/TldrawUiButtonCheck");
var import_TldrawUiButtonIcon = require("./lib/ui/components/primitives/Button/TldrawUiButtonIcon");
var import_TldrawUiButtonLabel = require("./lib/ui/components/primitives/Button/TldrawUiButtonLabel");
var import_TldrawUiMenuCheckboxItem = require("./lib/ui/components/primitives/menus/TldrawUiMenuCheckboxItem");
var import_TldrawUiMenuContext = require("./lib/ui/components/primitives/menus/TldrawUiMenuContext");
var import_TldrawUiMenuGroup = require("./lib/ui/components/primitives/menus/TldrawUiMenuGroup");
var import_TldrawUiMenuItem = require("./lib/ui/components/primitives/menus/TldrawUiMenuItem");
var import_TldrawUiMenuSubmenu = require("./lib/ui/components/primitives/menus/TldrawUiMenuSubmenu");
var import_TldrawUiContextualToolbar = require("./lib/ui/components/primitives/TldrawUiContextualToolbar");
var import_TldrawUiDialog = require("./lib/ui/components/primitives/TldrawUiDialog");
var import_TldrawUiDropdownMenu = require("./lib/ui/components/primitives/TldrawUiDropdownMenu");
var import_TldrawUiIcon = require("./lib/ui/components/primitives/TldrawUiIcon");
var import_TldrawUiInput = require("./lib/ui/components/primitives/TldrawUiInput");
var import_TldrawUiKbd = require("./lib/ui/components/primitives/TldrawUiKbd");
var import_TldrawUiPopover = require("./lib/ui/components/primitives/TldrawUiPopover");
var import_TldrawUiSelect = require("./lib/ui/components/primitives/TldrawUiSelect");
var import_TldrawUiSlider = require("./lib/ui/components/primitives/TldrawUiSlider");
var import_TldrawUiToolbar = require("./lib/ui/components/primitives/TldrawUiToolbar");
var import_TldrawUiTooltip = require("./lib/ui/components/primitives/TldrawUiTooltip");
var import_DefaultQuickActions = require("./lib/ui/components/QuickActions/DefaultQuickActions");
var import_DefaultQuickActionsContent = require("./lib/ui/components/QuickActions/DefaultQuickActionsContent");
var import_DefaultPeopleMenu = require("./lib/ui/components/SharePanel/DefaultPeopleMenu");
var import_DefaultPeopleMenuAvatar = require("./lib/ui/components/SharePanel/DefaultPeopleMenuAvatar");
var import_DefaultPeopleMenuContent = require("./lib/ui/components/SharePanel/DefaultPeopleMenuContent");
var import_DefaultPeopleMenuFacePile = require("./lib/ui/components/SharePanel/DefaultPeopleMenuFacePile");
var import_DefaultPeopleMenuItem = require("./lib/ui/components/SharePanel/DefaultPeopleMenuItem");
var import_DefaultSharePanel = require("./lib/ui/components/SharePanel/DefaultSharePanel");
var import_DefaultUserPresenceEditor = require("./lib/ui/components/SharePanel/DefaultUserPresenceEditor");
var import_Spinner = require("./lib/ui/components/Spinner");
var import_DefaultStylePanel = require("./lib/ui/components/StylePanel/DefaultStylePanel");
var import_DefaultStylePanelContent = require("./lib/ui/components/StylePanel/DefaultStylePanelContent");
var import_StylePanelButtonPicker = require("./lib/ui/components/StylePanel/StylePanelButtonPicker");
var import_StylePanelContext = require("./lib/ui/components/StylePanel/StylePanelContext");
var import_StylePanelDoubleDropdownPicker = require("./lib/ui/components/StylePanel/StylePanelDoubleDropdownPicker");
var import_StylePanelDropdownPicker = require("./lib/ui/components/StylePanel/StylePanelDropdownPicker");
var import_StylePanelSubheading = require("./lib/ui/components/StylePanel/StylePanelSubheading");
var import_DefaultImageToolbar = require("./lib/ui/components/Toolbar/DefaultImageToolbar");
var import_DefaultImageToolbarContent = require("./lib/ui/components/Toolbar/DefaultImageToolbarContent");
var import_DefaultRichTextToolbar = require("./lib/ui/components/Toolbar/DefaultRichTextToolbar");
var import_DefaultRichTextToolbarContent = require("./lib/ui/components/Toolbar/DefaultRichTextToolbarContent");
var import_DefaultToolbar = require("./lib/ui/components/Toolbar/DefaultToolbar");
var import_DefaultToolbarContent = require("./lib/ui/components/Toolbar/DefaultToolbarContent");
var import_DefaultVideoToolbar = require("./lib/ui/components/Toolbar/DefaultVideoToolbar");
var import_DefaultVideoToolbarContent = require("./lib/ui/components/Toolbar/DefaultVideoToolbarContent");
var import_OverflowingToolbar = require("./lib/ui/components/Toolbar/OverflowingToolbar");
var import_ToggleToolLockedButton = require("./lib/ui/components/Toolbar/ToggleToolLockedButton");
var import_CenteredTopPanelContainer = require("./lib/ui/components/TopPanel/CenteredTopPanelContainer");
var import_DefaultZoomMenu = require("./lib/ui/components/ZoomMenu/DefaultZoomMenu");
var import_DefaultZoomMenuContent = require("./lib/ui/components/ZoomMenu/DefaultZoomMenuContent");
var import_constants = require("./lib/ui/constants");
var import_a11y = require("./lib/ui/context/a11y");
var import_actions = require("./lib/ui/context/actions");
var import_asset_urls = require("./lib/ui/context/asset-urls");
var import_breakpoints = require("./lib/ui/context/breakpoints");
var import_components = require("./lib/ui/context/components");
var import_dialogs = require("./lib/ui/context/dialogs");
var import_events = require("./lib/ui/context/events");
var import_TldrawUiContextProvider = require("./lib/ui/context/TldrawUiContextProvider");
var import_toasts = require("./lib/ui/context/toasts");
var import_menu_hooks = require("./lib/ui/hooks/menu-hooks");
var import_useClipboardEvents = require("./lib/ui/hooks/useClipboardEvents");
var import_useCollaborationStatus = require("./lib/ui/hooks/useCollaborationStatus");
var import_useCopyAs = require("./lib/ui/hooks/useCopyAs");
var import_useExportAs = require("./lib/ui/hooks/useExportAs");
var import_useKeyboardShortcuts = require("./lib/ui/hooks/useKeyboardShortcuts");
var import_useLocalStorageState = require("./lib/ui/hooks/useLocalStorageState");
var import_useMenuIsOpen = require("./lib/ui/hooks/useMenuIsOpen");
var import_useReadonly = require("./lib/ui/hooks/useReadonly");
var import_useRelevantStyles = require("./lib/ui/hooks/useRelevantStyles");
var import_useTools = require("./lib/ui/hooks/useTools");
var import_translations = require("./lib/ui/hooks/useTranslation/translations");
var import_useTranslation2 = require("./lib/ui/hooks/useTranslation/useTranslation");
var import_icon_types = require("./lib/ui/icon-types");
var import_overrides = require("./lib/ui/overrides");
var import_TldrawUi = require("./lib/ui/TldrawUi");
var import_assets = require("./lib/utils/assets/assets");
var import_preload_font = require("./lib/utils/assets/preload-font");
var import_embeds = require("./lib/utils/embeds/embeds");
var import_putExcalidrawContent = require("./lib/utils/excalidraw/putExcalidrawContent");
var import_copyAs = require("./lib/utils/export/copyAs");
var import_exportAs = require("./lib/utils/export/exportAs");
var import_frames = require("./lib/utils/frames/frames");
var import_assetUrls2 = require("./lib/utils/static-assets/assetUrls");
var import_sanitizeSvg = require("./lib/utils/svg/sanitizeSvg");
var import_richText = require("./lib/utils/text/richText");
var import_text = require("./lib/utils/text/text");
var import_buildFromV1Document = require("./lib/utils/tldr/buildFromV1Document");
var import_file = require("./lib/utils/tldr/file");
(0, import_editor.registerTldrawLibraryVersion)(
  "tldraw",
  "5.2.4",
  "cjs"
);
