// @tldraw/tlschema/dist-cjs/records/TLAsset.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TLAsset_exports = {};
__export(TLAsset_exports, {
  AssetRecordType: () => AssetRecordType,
  assetMigrations: () => assetMigrations,
  assetVersions: () => assetVersions,
  createAssetPropsMigrationIds: () => createAssetPropsMigrationIds,
  createAssetPropsMigrationSequence: () => createAssetPropsMigrationSequence,
  createAssetRecordType: () => createAssetRecordType
});
module.exports = __toCommonJS(TLAsset_exports);
var import_store = require("@tldraw/store");
var import_utils = require("@tldraw/utils");
var import_validate = require("@tldraw/validate");
var import_TLBaseAsset = require("../assets/TLBaseAsset");
const assetVersions = (0, import_store.createMigrationIds)("com.tldraw.asset", {
  AddMeta: 1
});
const assetMigrations = (0, import_store.createRecordMigrationSequence)({
  sequenceId: "com.tldraw.asset",
  recordType: "asset",
  sequence: [
    {
      id: assetVersions.AddMeta,
      up: (record) => {
        ;
        record.meta = {};
      }
    }
  ]
});
function createAssetRecordType(assets) {
  return (0, import_store.createRecordType)("asset", {
    scope: "document",
    validator: import_validate.T.model(
      "asset",
      import_validate.T.union(
        "type",
        (0, import_utils.mapObjectMapValues)(
          assets,
          (type, { props, meta }) => (0, import_TLBaseAsset.createAssetValidator)(type, props, meta)
        )
      )
    )
  }).withDefaultProperties(() => ({
    meta: {}
  }));
}
const AssetRecordType = (0, import_store.createRecordType)("asset", {
  scope: "document"
}).withDefaultProperties(() => ({
  meta: {}
}));
function createAssetPropsMigrationSequence(migrations) {
  return migrations;
}
function createAssetPropsMigrationIds(assetType, ids) {
  return (0, import_utils.mapObjectMapValues)(ids, (_k, v) => `com.tldraw.asset.${assetType}/${v}`);
}
