// @tldraw/tlschema/dist-cjs/records/TLUser.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var TLUser_exports = {};
__export(TLUser_exports, {
  UserRecordType: () => UserRecordType,
  createUserId: () => createUserId,
  createUserRecordType: () => createUserRecordType,
  isUserId: () => isUserId,
  userIdValidator: () => userIdValidator,
  userMigrations: () => userMigrations,
  userValidator: () => userValidator,
  userVersions: () => userVersions
});
module.exports = __toCommonJS(TLUser_exports);
var import_store = require("@tldraw/store");
var import_validate = require("@tldraw/validate");
var import_id_validator = require("../misc/id-validator");
const userIdValidator = (0, import_id_validator.idValidator)("user");
const userVersions = (0, import_store.createMigrationIds)("com.tldraw.user", {
  Initial: 1
});
const userMigrations = (0, import_store.createRecordMigrationSequence)({
  sequenceId: "com.tldraw.user",
  recordType: "user",
  sequence: [
    {
      id: userVersions.Initial,
      up: (_record) => {
      }
    }
  ]
});
function createUserRecordType(config) {
  const metaConfig = config?.meta;
  const metaValidator = metaConfig ? import_validate.T.object(
    Object.fromEntries(
      Object.entries(metaConfig).map(([key, v]) => [
        key,
        new import_validate.T.Validator((value) => {
          if (value === void 0) return void 0;
          return v.validate(value);
        })
      ])
    )
  ).allowUnknownProperties() : import_validate.T.jsonValue;
  const validator = import_validate.T.model(
    "user",
    import_validate.T.object({
      typeName: import_validate.T.literal("user"),
      id: userIdValidator,
      name: import_validate.T.string,
      color: import_validate.T.string,
      imageUrl: import_validate.T.string,
      meta: metaValidator
    })
  );
  return (0, import_store.createRecordType)("user", {
    validator,
    scope: "document"
  }).withDefaultProperties(() => ({
    name: "",
    color: "",
    imageUrl: "",
    meta: {}
  }));
}
const userValidator = import_validate.T.model(
  "user",
  import_validate.T.object({
    typeName: import_validate.T.literal("user"),
    id: userIdValidator,
    name: import_validate.T.string,
    color: import_validate.T.string,
    imageUrl: import_validate.T.string,
    meta: import_validate.T.jsonValue
  })
);
const UserRecordType = createUserRecordType();
function isUserId(id) {
  return UserRecordType.isId(id);
}
function createUserId(id) {
  return UserRecordType.createId(id);
}
