// @tldraw/editor/dist-cjs/lib/utils/runtime.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var runtime_exports = {};
__export(runtime_exports, {
  hardResetEditor: () => hardResetEditor,
  openWindow: () => openWindow,
  refreshPage: () => refreshPage,
  runtime: () => runtime,
  setRuntimeOverrides: () => setRuntimeOverrides
});
module.exports = __toCommonJS(runtime_exports);
var import_hardReset = require("./sync/hardReset");
const runtime = {
  openWindow(url, target, allowReferrer = false) {
    return window.open(url, target, allowReferrer ? "noopener" : "noopener noreferrer");
  },
  refreshPage() {
    window.location.reload();
  },
  async hardReset() {
    return await (0, import_hardReset.hardReset)({ shouldReload: true });
  }
};
function setRuntimeOverrides(input) {
  Object.assign(runtime, input);
}
function openWindow(url, target = "_blank", allowReferrer) {
  return runtime.openWindow(url, target, allowReferrer);
}
function refreshPage() {
  runtime.refreshPage();
}
function hardResetEditor() {
  runtime.hardReset();
}
